// @trigger.dev/core/dist/commonjs/v3/waitUntil/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WaitUntilAPI = void 0;
const globals_js_1 = require("../utils/globals.js");
const API_NAME = "wait-until";
class NoopManager {
    register(promise) {
        // noop
    }
    blockUntilSettled(timeout) {
        return Promise.resolve();
    }
    requiresResolving() {
        return false;
    }
}
const NOOP_MANAGER = new NoopManager();
class WaitUntilAPI {
    static _instance;
    constructor() { }
    static getInstance() {
        if (!this._instance) {
            this._instance = new WaitUntilAPI();
        }
        return this._instance;
    }
    setGlobalManager(manager) {
        return (0, globals_js_1.registerGlobal)(API_NAME, manager);
    }
    #getManager() {
        return (0, globals_js_1.getGlobal)(API_NAME) ?? NOOP_MANAGER;
    }
    register(promise) {
        return this.#getManager().register(promise);
    }
    blockUntilSettled(timeout) {
        return this.#getManager().blockUntilSettled(timeout);
    }
    requiresResolving() {
        return this.#getManager().requiresResolving();
    }
}
exports.WaitUntilAPI = WaitUntilAPI;
