// @tanstack/start-api-routes/dist/esm/index.js
import { eventHandler, toWebRequest } from "@tanstack/start-server-core";
import vinxiFileRoutes from "vinxi/routes";
const HTTP_API_METHODS = [
  "GET",
  "POST",
  "PUT",
  "PATCH",
  "DELETE",
  "OPTIONS",
  "HEAD"
];
function createStartAPIHandler(cb) {
  return eventHandler(async (event) => {
    const request = toWebRequest(event);
    const res = await cb({ request });
    return res;
  });
}
const createAPIRoute = (path) => (methods) => ({
  path,
  methods
});
const createAPIFileRoute = (filePath) => (methods) => ({
  path: filePath,
  methods
});
function findRoute(url, entryRoutes) {
  const urlSegments = url.pathname.split("/").filter(Boolean);
  const routes = entryRoutes.sort((a, b) => {
    const aParts = a.routePath.split("/").filter(Boolean);
    const bParts = b.routePath.split("/").filter(Boolean);
    return bParts.length - aParts.length;
  }).filter((r) => {
    const routeSegments = r.routePath.split("/").filter(Boolean);
    return urlSegments.length >= routeSegments.length;
  });
  for (const route of routes) {
    const routeSegments = route.routePath.split("/").filter(Boolean);
    const params = {};
    let matches = true;
    for (let i = 0; i < routeSegments.length; i++) {
      const routeSegment = routeSegments[i];
      const urlSegment = urlSegments[i];
      if (routeSegment.startsWith("$")) {
        if (routeSegment === "$") {
          const wildcardValue = urlSegments.slice(i).join("/");
          if (wildcardValue !== "") {
            params["*"] = wildcardValue;
            params["_splat"] = wildcardValue;
          } else {
            matches = false;
            break;
          }
        } else {
          const paramName = routeSegment.slice(1);
          params[paramName] = urlSegment;
        }
      } else if (routeSegment !== urlSegment) {
        matches = false;
        break;
      }
    }
    if (matches) {
      return { routePath: route.routePath, params, payload: route.payload };
    }
  }
  return void 0;
}
const defaultAPIRoutesHandler = (opts) => {
  return async ({ request }) => {
    if (!HTTP_API_METHODS.includes(request.method)) {
      return new Response("Method not allowed", { status: 405 });
    }
    const url = new URL(request.url, "http://localhost:3000");
    const routes = Object.entries(opts.routes).map(([routePath, route]) => ({
      routePath,
      payload: route
    }));
    const match = findRoute(url, routes);
    if (!match) {
      return new Response("Not found", { status: 404 });
    }
    if (match.routePath !== match.payload.path) {
      console.error(
        `Route path mismatch: ${match.routePath} !== ${match.payload.path}. Please make sure that the route path in \`createAPIRoute\` matches the path in the handler map in \`defaultAPIRoutesHandler\``
      );
      return new Response("Not found", { status: 404 });
    }
    const method = request.method;
    const handler = match.payload.methods[method];
    if (!handler) {
      return new Response("Method not allowed", { status: 405 });
    }
    return await handler({ request, params: match.params });
  };
};
const vinxiRoutes = vinxiFileRoutes.filter((route) => route["$APIRoute"]);
function toTSRFileBasedRoutes(routes) {
  const pairs = [];
  routes.forEach((route) => {
    const parts = route.path.split("/").filter(Boolean);
    const path = parts.map((part) => {
      if (part === "*splat") {
        return "$";
      }
      if (part.startsWith(":$") && part.endsWith("?")) {
        return part.slice(1, -1);
      }
      return part;
    }).join("/");
    pairs.push({ routePath: `/${path}`, payload: route });
  });
  return pairs;
}
const defaultAPIFileRouteHandler = async ({
  request
}) => {
  if (!vinxiRoutes.length) {
    return new Response("No routes found", { status: 404 });
  }
  if (!HTTP_API_METHODS.includes(request.method)) {
    return new Response("Method not allowed", { status: 405 });
  }
  const routes = toTSRFileBasedRoutes(vinxiRoutes);
  const url = new URL(request.url, "http://localhost:3000");
  const match = findRoute(url, routes);
  if (!match) {
    return new Response("Not found", { status: 404 });
  }
  let action = void 0;
  try {
    action = await match.payload.$APIRoute.import().then((m) => m.APIRoute);
  } catch (err) {
    console.error("Error importing route file:", err);
    return new Response("Internal server error", { status: 500 });
  }
  if (!action) {
    return new Response("Internal server error", { status: 500 });
  }
  const method = request.method;
  const handler = action.methods[method];
  if (!handler) {
    return new Response("Method not allowed", { status: 405 });
  }
  return await handler({ request, params: match.params });
};
export {
  createAPIFileRoute,
  createAPIRoute,
  createStartAPIHandler,
  defaultAPIFileRouteHandler,
  defaultAPIRoutesHandler
};
