// @tanstack/directive-functions-plugin/dist/cjs/compilers.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const babel = require("@babel/core");
const types = require("@babel/types");
const codeFrame = require("@babel/code-frame");
const babelDeadCodeElimination = require("babel-dead-code-elimination");
const path = require("pathe");
const routerUtils = require("@tanstack/router-utils");
function _interopNamespaceDefault(e) {
  const n = Object.create(null, { [Symbol.toStringTag]: { value: "Module" } });
  if (e) {
    for (const k in e) {
      if (k !== "default") {
        const d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: () => e[k]
        });
      }
    }
  }
  n.default = e;
  return Object.freeze(n);
}
const babel__namespace = /* @__PURE__ */ _interopNamespaceDefault(babel);
function compileDirectives(opts) {
  const ast = routerUtils.parseAst(opts);
  const refIdents = babelDeadCodeElimination.findReferencedIdentifiers(ast);
  const directiveFnsById = findDirectives(ast, {
    ...opts,
    directiveSplitParam: opts.directiveSplitParam,
    isDirectiveSplitParam: opts.isDirectiveSplitParam
  });
  if (Object.keys(directiveFnsById).length > 0) {
    if (opts.getRuntimeCode) {
      const runtimeImport = babel__namespace.template.statement(
        opts.getRuntimeCode({ directiveFnsById })
      )();
      ast.program.body.unshift(runtimeImport);
    }
  }
  if (opts.isDirectiveSplitParam) {
    safeRemoveExports(ast);
    ast.program.body.push(
      babel__namespace.types.exportNamedDeclaration(
        void 0,
        Object.values(directiveFnsById).map(
          (fn) => babel__namespace.types.exportSpecifier(
            babel__namespace.types.identifier(fn.functionName),
            babel__namespace.types.identifier(fn.functionName)
          )
        )
      )
    );
  }
  babelDeadCodeElimination.deadCodeElimination(ast, refIdents);
  const compiledResult = routerUtils.generateFromAst(ast, {
    sourceMaps: true,
    sourceFileName: opts.filename,
    filename: opts.filename
  });
  return {
    compiledResult,
    directiveFnsById
  };
}
function findNearestVariableName(path2, directive) {
  let currentPath = path2;
  const nameParts = [];
  while (currentPath) {
    const name = (() => {
      if (babel__namespace.types.isFunctionExpression(currentPath.node) && currentPath.node.id) {
        return currentPath.node.id.name;
      }
      if (babel__namespace.types.isCallExpression(currentPath.node)) {
        const current = currentPath.node.callee;
        const chainParts = [];
        if (babel__namespace.types.isMemberExpression(current)) {
          if (babel__namespace.types.isIdentifier(current.property)) {
            chainParts.unshift(current.property.name);
          }
          let base = current.object;
          while (!babel__namespace.types.isIdentifier(base)) {
            if (babel__namespace.types.isCallExpression(base)) {
              base = base.callee;
            } else if (babel__namespace.types.isMemberExpression(base)) {
              base = base.object;
            } else {
              break;
            }
          }
          if (babel__namespace.types.isIdentifier(base)) {
            chainParts.unshift(base.name);
          }
        } else if (babel__namespace.types.isIdentifier(current)) {
          chainParts.unshift(current.name);
        }
        if (chainParts.length > 0) {
          return chainParts.join("_");
        }
      }
      if (babel__namespace.types.isFunctionDeclaration(currentPath.node)) {
        return currentPath.node.id?.name;
      }
      if (babel__namespace.types.isIdentifier(currentPath.node)) {
        return currentPath.node.name;
      }
      if (types.isVariableDeclarator(currentPath.node) && types.isIdentifier(currentPath.node.id)) {
        return currentPath.node.id.name;
      }
      if (babel__namespace.types.isClassMethod(currentPath.node) || babel__namespace.types.isObjectMethod(currentPath.node)) {
        throw new Error(
          `"${directive}" in ClassMethod or ObjectMethod not supported`
        );
      }
      return "";
    })();
    if (name) {
      nameParts.unshift(name);
    }
    currentPath = currentPath.parentPath;
  }
  return nameParts.length > 0 ? nameParts.join("_") : "anonymous";
}
function makeIdentifierSafe(identifier) {
  return identifier.replace(/[^a-zA-Z0-9_$]/g, "_").replace(/^[0-9]/, "_$&").replace(/^\$/, "_$").replace(/_{2,}/g, "_").replace(/^_|_$/g, "");
}
function findDirectives(ast, opts) {
  const directiveFnsById = {};
  const functionNameSet = /* @__PURE__ */ new Set();
  let programPath;
  babel__namespace.traverse(ast, {
    Program(path2) {
      programPath = path2;
    }
  });
  const hasFileDirective = ast.program.directives.some(
    (directive) => directive.value.value === opts.directive
  );
  const compileDirectiveOpts = {
    isDirectiveSplitParam: opts.isDirectiveSplitParam
  };
  if (hasFileDirective) {
    babel__namespace.traverse(ast, {
      ExportDefaultDeclaration(path2) {
        if (babel__namespace.types.isFunctionDeclaration(path2.node.declaration)) {
          compileDirective(
            path2.get("declaration"),
            compileDirectiveOpts
          );
        }
      },
      ExportNamedDeclaration(path2) {
        if (babel__namespace.types.isFunctionDeclaration(path2.node.declaration)) {
          compileDirective(
            path2.get("declaration"),
            compileDirectiveOpts
          );
        }
      },
      ExportDeclaration(path2) {
        if (babel__namespace.types.isExportNamedDeclaration(path2.node) && babel__namespace.types.isVariableDeclaration(path2.node.declaration) && (babel__namespace.types.isFunctionExpression(
          path2.node.declaration.declarations[0]?.init
        ) || babel__namespace.types.isArrowFunctionExpression(
          path2.node.declaration.declarations[0]?.init
        ))) {
          compileDirective(
            path2.get(
              "declaration.declarations.0.init"
            ),
            compileDirectiveOpts
          );
        }
      }
    });
  } else {
    babel__namespace.traverse(ast, {
      DirectiveLiteral(nodePath) {
        if (nodePath.node.value === opts.directive) {
          const directiveFn = nodePath.findParent((p) => p.isFunction());
          if (!directiveFn) return;
          const isClassMethod = directiveFn.isClassMethod();
          const isObjectMethod = directiveFn.isObjectMethod();
          if (isClassMethod || isObjectMethod) {
            throw codeFrameError(
              opts.code,
              directiveFn.node.loc,
              `"${opts.directive}" in ${isClassMethod ? "class" : isObjectMethod ? "object method" : ""} not supported`
            );
          }
          const nearestBlock = directiveFn.findParent(
            (p) => (p.isBlockStatement() || p.isScopable()) && !p.isProgram()
          );
          if (nearestBlock) {
            throw codeFrameError(
              opts.code,
              nearestBlock.node.loc,
              `"${opts.directive}" cannot be nested in other blocks or functions`
            );
          }
          if (!directiveFn.isFunctionDeclaration() && !directiveFn.isFunctionExpression() && !(directiveFn.isArrowFunctionExpression() && babel__namespace.types.isBlockStatement(directiveFn.node.body))) {
            throw codeFrameError(
              opts.code,
              directiveFn.node.loc,
              `"${opts.directive}" must be function declarations or function expressions`
            );
          }
          compileDirective(directiveFn, compileDirectiveOpts);
        }
      }
    });
  }
  return directiveFnsById;
  function compileDirective(directiveFn, compileDirectiveOpts2) {
    const programBody = programPath.node.body;
    if (babel__namespace.types.isFunction(directiveFn.node) && babel__namespace.types.isBlockStatement(directiveFn.node.body)) {
      directiveFn.node.body.directives = directiveFn.node.body.directives.filter(
        (directive) => directive.value.value !== opts.directive
      );
    }
    if (directiveFn.parentPath.isProgram()) {
      if (!babel__namespace.types.isFunctionDeclaration(directiveFn.node)) {
        throw new Error("Top level functions must be function declarations");
      }
      const index = programBody.indexOf(directiveFn.node);
      const originalFunctionName = directiveFn.node.id.name;
      directiveFn.node.id = null;
      const variableDeclaration = babel__namespace.types.variableDeclaration("const", [
        babel__namespace.types.variableDeclarator(
          babel__namespace.types.identifier(originalFunctionName),
          babel__namespace.types.toExpression(directiveFn.node)
        )
      ]);
      directiveFn.replaceWith(variableDeclaration);
      directiveFn = programPath.get(
        `body.${index}.declarations.0.init`
      );
    }
    let functionName = findNearestVariableName(directiveFn, opts.directive);
    const incrementFunctionNameVersion = (functionName2) => {
      const [realReferenceName, count] = functionName2.split(/_(\d+)$/);
      const resolvedCount = Number(count || "0");
      const suffix = `_${resolvedCount + 1}`;
      return makeIdentifierSafe(realReferenceName) + suffix;
    };
    while (functionNameSet.has(functionName)) {
      functionName = incrementFunctionNameVersion(functionName);
    }
    functionNameSet.add(functionName);
    while (programPath.scope.hasBinding(functionName)) {
      functionName = incrementFunctionNameVersion(functionName);
      programPath.scope.crawl();
    }
    functionNameSet.add(functionName);
    const [baseFilename, ..._searchParams] = opts.filename.split("?");
    const searchParams = new URLSearchParams(_searchParams.join("&"));
    searchParams.set(opts.directiveSplitParam, "");
    const localExtractedFilename = `${baseFilename}?${searchParams.toString()}`;
    const relativeFilename = path.relative(opts.root, baseFilename);
    const functionId = opts.generateFunctionId({
      filename: relativeFilename,
      functionName,
      extractedFilename: localExtractedFilename
    });
    const knownFn = opts.knownDirectiveFns?.[functionId];
    const extractedFilename = knownFn?.extractedFilename ?? localExtractedFilename;
    const isClientReferenced = opts.isClientEnvironment || !!knownFn;
    if (isClientReferenced && !compileDirectiveOpts2.isDirectiveSplitParam) {
      let currentPath = directiveFn;
      while (currentPath && !currentPath.parentPath?.isProgram()) {
        const parent = currentPath.parentPath;
        if (parent?.isCallExpression() && babel__namespace.types.isMemberExpression(parent.node.callee) && babel__namespace.types.isIdentifier(parent.node.callee.property) && parent.node.callee.property.name === "handler" && parent.node.arguments.length === 2) {
          parent.node.arguments.pop();
          break;
        }
        currentPath = parent;
      }
    }
    const topParent = directiveFn.findParent((p) => !!p.parentPath?.isProgram()) || directiveFn;
    const topParentIndex = programBody.indexOf(topParent.node);
    if (directiveFn.parentPath.isProgram()) {
      throw new Error(
        "Top level functions should have already been compiled to variable declarations by this point"
      );
    }
    programBody.splice(
      topParentIndex,
      0,
      babel__namespace.types.variableDeclaration("const", [
        babel__namespace.types.variableDeclarator(
          babel__namespace.types.identifier(functionName),
          babel__namespace.types.toExpression(directiveFn.node)
        )
      ])
    );
    if ((babel__namespace.types.isExportNamedDeclaration(directiveFn.parentPath.node) || compileDirectiveOpts2.isDirectiveSplitParam && babel__namespace.types.isExportDefaultDeclaration(
      directiveFn.parentPath.node
    )) && (babel__namespace.types.isFunctionDeclaration(directiveFn.node) || babel__namespace.types.isFunctionExpression(directiveFn.node)) && babel__namespace.types.isIdentifier(directiveFn.node.id)) {
      const originalFunctionName = directiveFn.node.id.name;
      programBody.splice(
        topParentIndex + 1,
        0,
        babel__namespace.types.exportNamedDeclaration(
          babel__namespace.types.variableDeclaration("const", [
            babel__namespace.types.variableDeclarator(
              babel__namespace.types.identifier(originalFunctionName),
              babel__namespace.types.identifier(functionName)
            )
          ])
        )
      );
      directiveFn.remove();
    } else {
      directiveFn.replaceWith(babel__namespace.types.identifier(functionName));
    }
    directiveFn = programPath.get(
      `body.${topParentIndex}.declarations.0.init`
    );
    if (opts.replacer) {
      const replacer = opts.replacer({
        fn: "$$fn$$",
        extractedFilename,
        filename: opts.filename,
        functionId,
        functionName,
        isSourceFn: !!opts.directiveSplitParam,
        isClientReferenced
      });
      const replacement = babel__namespace.template.expression(replacer, {
        placeholderPattern: false,
        placeholderWhitelist: /* @__PURE__ */ new Set(["$$fn$$"])
      })({
        ...replacer.includes("$$fn$$") ? { $$fn$$: babel__namespace.types.toExpression(directiveFn.node) } : {}
      });
      directiveFn.replaceWith(replacement);
    }
    directiveFnsById[functionId] = {
      nodePath: directiveFn,
      functionName,
      functionId,
      extractedFilename,
      filename: opts.filename,
      chunkName: fileNameToChunkName(opts.root, extractedFilename),
      isClientReferenced
    };
  }
}
function codeFrameError(code, loc, message) {
  if (!loc) {
    return new Error(`${message} at unknown location`);
  }
  const frame = codeFrame.codeFrameColumns(
    code,
    {
      start: loc.start,
      end: loc.end
    },
    {
      highlightCode: true,
      message
    }
  );
  return new Error(frame);
}
const safeRemoveExports = (ast) => {
  ast.program.body = ast.program.body.flatMap((node) => {
    if (babel__namespace.types.isExportNamedDeclaration(node) || babel__namespace.types.isExportDefaultDeclaration(node)) {
      if (babel__namespace.types.isFunctionDeclaration(node.declaration) || babel__namespace.types.isClassDeclaration(node.declaration) || babel__namespace.types.isVariableDeclaration(node.declaration)) {
        if (babel__namespace.types.isFunctionDeclaration(node.declaration) || babel__namespace.types.isClassDeclaration(node.declaration)) {
          if (!node.declaration.id) {
            return node;
          }
        }
        return node.declaration;
      } else if (node.declaration === null) {
        return [];
      }
    }
    return node;
  });
};
function fileNameToChunkName(root, fileName) {
  return fileName.replace(root, "").replace(/[^a-zA-Z0-9_]/g, "_");
}
exports.compileDirectives = compileDirectives;
exports.findDirectives = findDirectives;
