// @tanstack/directive-functions-plugin/dist/cjs/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const node_url = require("node:url");
const routerUtils = require("@tanstack/router-utils");
const compilers = require("./compilers.cjs");
const debug = process.env.TSR_VITE_DEBUG && ["true", "directives-functions-plugin"].includes(process.env.TSR_VITE_DEBUG);
const createDirectiveRx = (directive) => new RegExp(`"${directive}"|'${directive}'`, "gm");
function buildDirectiveSplitParam(directive) {
  return `tsr-directive-${directive.replace(/[^a-zA-Z0-9]/g, "-")}`;
}
function TanStackDirectiveFunctionsPluginEnv(opts) {
  let root = process.cwd();
  const directiveRx = createDirectiveRx(opts.directive);
  const appliedEnvironments = /* @__PURE__ */ new Set([
    ...opts.callers.map((c) => c.envName),
    opts.provider.envName
  ]);
  const directiveSplitParam = buildDirectiveSplitParam(opts.directive);
  return {
    name: "tanstack-start-directive-vite-plugin",
    enforce: "pre",
    buildStart() {
      root = this.environment.config.root;
    },
    applyToEnvironment(env) {
      return appliedEnvironments.has(env.name);
    },
    transform: {
      filter: {
        code: directiveRx
      },
      handler(code, id) {
        const url = node_url.pathToFileURL(id);
        url.searchParams.delete("v");
        id = node_url.fileURLToPath(url).replace(/\\/g, "/");
        const isDirectiveSplitParam = id.includes(directiveSplitParam);
        let envOptions;
        if (isDirectiveSplitParam) {
          envOptions = opts.provider;
          if (debug)
            console.info(
              `Compiling Directives for provider in environment ${envOptions.envName}: `,
              id
            );
        } else {
          const callerOptions = opts.callers.find(
            (e) => e.envName === this.environment.name
          );
          if (callerOptions) {
            envOptions = callerOptions;
            if (debug)
              console.info(
                `Compiling Directives for caller in environment ${envOptions.envName}: `,
                id
              );
          } else {
            envOptions = opts.provider;
            if (debug)
              console.info(
                `Compiling Directives for provider (fallback) in environment ${opts.provider.envName}: `,
                id
              );
          }
        }
        const knownDirectiveFns = opts.getKnownDirectiveFns?.();
        const isClientEnvironment = envOptions.envConsumer === "client";
        const { compiledResult, directiveFnsById } = compilers.compileDirectives({
          directive: opts.directive,
          getRuntimeCode: envOptions.getRuntimeCode,
          generateFunctionId: opts.generateFunctionId,
          replacer: envOptions.replacer,
          code,
          root,
          filename: id,
          directiveSplitParam,
          isDirectiveSplitParam,
          knownDirectiveFns,
          isClientEnvironment
        });
        if (!isDirectiveSplitParam) {
          opts.onDirectiveFnsById?.(directiveFnsById);
        }
        if (debug) {
          routerUtils.logDiff(code, compiledResult.code);
          console.log("Output:\n", compiledResult.code + "\n\n");
        }
        return compiledResult;
      }
    }
  };
}
exports.TanStackDirectiveFunctionsPluginEnv = TanStackDirectiveFunctionsPluginEnv;
