// @pulumi/cloudflare/accessPolicy.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccessPolicy = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Access: Apps and Policies Read`
 * - `Access: Apps and Policies Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleZeroTrustAccessPolicy = new cloudflare.ZeroTrustAccessPolicy("example_zero_trust_access_policy", {
 *     accountId: "023e105f4ecef8ad9ca31a8372d0c353",
 *     decision: "allow",
 *     includes: [{
 *         certificate: {},
 *     }],
 *     name: "Allow devs",
 *     approvalGroups: [
 *         {
 *             approvalsNeeded: 1,
 *             emailAddresses: [
 *                 "test1@cloudflare.com",
 *                 "test2@cloudflare.com",
 *             ],
 *             emailListUuid: "email_list_uuid",
 *         },
 *         {
 *             approvalsNeeded: 3,
 *             emailAddresses: [
 *                 "test@cloudflare.com",
 *                 "test2@cloudflare.com",
 *             ],
 *             emailListUuid: "597147a1-976b-4ef2-9af0-81d5d007fc34",
 *         },
 *     ],
 *     approvalRequired: true,
 *     connectionRules: {
 *         rdp: {
 *             allowedClipboardLocalToRemoteFormats: ["text"],
 *             allowedClipboardRemoteToLocalFormats: ["text"],
 *         },
 *     },
 *     excludes: [{
 *         certificate: {},
 *     }],
 *     isolationRequired: false,
 *     mfaConfig: {
 *         allowedAuthenticators: [
 *             "totp",
 *             "biometrics",
 *             "security_key",
 *         ],
 *         mfaDisabled: false,
 *         sessionDuration: "24h",
 *     },
 *     purposeJustificationPrompt: "Please enter a justification for entering this protected domain.",
 *     purposeJustificationRequired: true,
 *     requires: [{
 *         certificate: {},
 *     }],
 *     sessionDuration: "24h",
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/accessPolicy:AccessPolicy example '<account_id>/<policy_id>'
 * ```
 *
 * @deprecated cloudflare.index/accesspolicy.AccessPolicy has been deprecated in favor of cloudflare.index/zerotrustaccesspolicy.ZeroTrustAccessPolicy
 */
class AccessPolicy extends pulumi.CustomResource {
    /**
     * Get an existing AccessPolicy resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        pulumi.log.warn("AccessPolicy is deprecated: cloudflare.index/accesspolicy.AccessPolicy has been deprecated in favor of cloudflare.index/zerotrustaccesspolicy.ZeroTrustAccessPolicy");
        return new AccessPolicy(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/accessPolicy:AccessPolicy';
    /**
     * Returns true if the given object is an instance of AccessPolicy.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === AccessPolicy.__pulumiType;
    }
    /** @deprecated cloudflare.index/accesspolicy.AccessPolicy has been deprecated in favor of cloudflare.index/zerotrustaccesspolicy.ZeroTrustAccessPolicy */
    constructor(name, argsOrState, opts) {
        pulumi.log.warn("AccessPolicy is deprecated: cloudflare.index/accesspolicy.AccessPolicy has been deprecated in favor of cloudflare.index/zerotrustaccesspolicy.ZeroTrustAccessPolicy");
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["approvalGroups"] = state?.approvalGroups;
            resourceInputs["approvalRequired"] = state?.approvalRequired;
            resourceInputs["connectionRules"] = state?.connectionRules;
            resourceInputs["decision"] = state?.decision;
            resourceInputs["excludes"] = state?.excludes;
            resourceInputs["includes"] = state?.includes;
            resourceInputs["isolationRequired"] = state?.isolationRequired;
            resourceInputs["mfaConfig"] = state?.mfaConfig;
            resourceInputs["name"] = state?.name;
            resourceInputs["purposeJustificationPrompt"] = state?.purposeJustificationPrompt;
            resourceInputs["purposeJustificationRequired"] = state?.purposeJustificationRequired;
            resourceInputs["requires"] = state?.requires;
            resourceInputs["sessionDuration"] = state?.sessionDuration;
        }
        else {
            const args = argsOrState;
            if (args?.decision === undefined && !opts.urn) {
                throw new Error("Missing required property 'decision'");
            }
            if (args?.name === undefined && !opts.urn) {
                throw new Error("Missing required property 'name'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["approvalGroups"] = args?.approvalGroups;
            resourceInputs["approvalRequired"] = args?.approvalRequired;
            resourceInputs["connectionRules"] = args?.connectionRules;
            resourceInputs["decision"] = args?.decision;
            resourceInputs["excludes"] = args?.excludes;
            resourceInputs["includes"] = args?.includes;
            resourceInputs["isolationRequired"] = args?.isolationRequired;
            resourceInputs["mfaConfig"] = args?.mfaConfig;
            resourceInputs["name"] = args?.name;
            resourceInputs["purposeJustificationPrompt"] = args?.purposeJustificationPrompt;
            resourceInputs["purposeJustificationRequired"] = args?.purposeJustificationRequired;
            resourceInputs["requires"] = args?.requires;
            resourceInputs["sessionDuration"] = args?.sessionDuration;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const aliasOpts = { aliases: [{ type: "cloudflare:index/accessPolicy:AccessPolicy" }] };
        opts = pulumi.mergeOptions(opts, aliasOpts);
        super(AccessPolicy.__pulumiType, name, resourceInputs, opts);
    }
}
exports.AccessPolicy = AccessPolicy;
