// @pulumi/cloudflare/accountToken.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountToken = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Account API Tokens Read`
 * - `Account API Tokens Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleAccountToken = new cloudflare.AccountToken("example_account_token", {
 *     accountId: "b67e14daa5f8dceeb91fe5449ba496eb",
 *     name: "workers read-only token",
 *     policies: [{
 *         effect: "allow",
 *         permissionGroups: [
 *             {
 *                 id: "1a71c399035b4950a1bd1466bbe4f420",
 *             },
 *             {
 *                 id: "8b47d2786a534c08a1f94ee8f9f599ef",
 *             },
 *         ],
 *         resources: JSON.stringify({
 *             "com.cloudflare.api.account.b67e14daa5f8dceeb91fe5449ba496eb": "*",
 *         }),
 *     }],
 *     condition: {
 *         requestIp: {
 *             ins: [
 *                 "123.123.123.0/24",
 *                 "2606:4700::/32",
 *             ],
 *             notIns: [
 *                 "123.123.123.0/28",
 *                 "2606:4700:4700::/48",
 *             ],
 *         },
 *     },
 *     expiresOn: "2027-10-01T00:00:00Z",
 *     notBefore: "2025-10-01T00:00:00Z",
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/accountToken:AccountToken example '<account_id>/<token_id>'
 * ```
 */
class AccountToken extends pulumi.CustomResource {
    /**
     * Get an existing AccountToken resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new AccountToken(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/accountToken:AccountToken';
    /**
     * Returns true if the given object is an instance of AccountToken.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === AccountToken.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["condition"] = state?.condition;
            resourceInputs["expiresOn"] = state?.expiresOn;
            resourceInputs["issuedOn"] = state?.issuedOn;
            resourceInputs["lastUsedOn"] = state?.lastUsedOn;
            resourceInputs["modifiedOn"] = state?.modifiedOn;
            resourceInputs["name"] = state?.name;
            resourceInputs["notBefore"] = state?.notBefore;
            resourceInputs["policies"] = state?.policies;
            resourceInputs["status"] = state?.status;
            resourceInputs["value"] = state?.value;
        }
        else {
            const args = argsOrState;
            if (args?.accountId === undefined && !opts.urn) {
                throw new Error("Missing required property 'accountId'");
            }
            if (args?.name === undefined && !opts.urn) {
                throw new Error("Missing required property 'name'");
            }
            if (args?.policies === undefined && !opts.urn) {
                throw new Error("Missing required property 'policies'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["condition"] = args?.condition;
            resourceInputs["expiresOn"] = args?.expiresOn;
            resourceInputs["name"] = args?.name;
            resourceInputs["notBefore"] = args?.notBefore;
            resourceInputs["policies"] = args?.policies;
            resourceInputs["status"] = args?.status;
            resourceInputs["issuedOn"] = undefined /*out*/;
            resourceInputs["lastUsedOn"] = undefined /*out*/;
            resourceInputs["modifiedOn"] = undefined /*out*/;
            resourceInputs["value"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const secretOpts = { additionalSecretOutputs: ["value"] };
        opts = pulumi.mergeOptions(opts, secretOpts);
        super(AccountToken.__pulumiType, name, resourceInputs, opts);
    }
}
exports.AccountToken = AccountToken;
