// @pulumi/cloudflare/aiSearchInstance.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AiSearchInstance = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * ## Import
 *
 * > This resource does not currently support `pulumi import`.
 */
class AiSearchInstance extends pulumi.CustomResource {
    /**
     * Get an existing AiSearchInstance resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new AiSearchInstance(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/aiSearchInstance:AiSearchInstance';
    /**
     * Returns true if the given object is an instance of AiSearchInstance.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === AiSearchInstance.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["aiGatewayId"] = state?.aiGatewayId;
            resourceInputs["aiSearchInstanceId"] = state?.aiSearchInstanceId;
            resourceInputs["aisearchModel"] = state?.aisearchModel;
            resourceInputs["cache"] = state?.cache;
            resourceInputs["cacheThreshold"] = state?.cacheThreshold;
            resourceInputs["chunk"] = state?.chunk;
            resourceInputs["chunkOverlap"] = state?.chunkOverlap;
            resourceInputs["chunkSize"] = state?.chunkSize;
            resourceInputs["createdAt"] = state?.createdAt;
            resourceInputs["createdBy"] = state?.createdBy;
            resourceInputs["customMetadatas"] = state?.customMetadatas;
            resourceInputs["embeddingModel"] = state?.embeddingModel;
            resourceInputs["enable"] = state?.enable;
            resourceInputs["engineVersion"] = state?.engineVersion;
            resourceInputs["fusionMethod"] = state?.fusionMethod;
            resourceInputs["hybridSearchEnabled"] = state?.hybridSearchEnabled;
            resourceInputs["indexMethod"] = state?.indexMethod;
            resourceInputs["indexingOptions"] = state?.indexingOptions;
            resourceInputs["lastActivity"] = state?.lastActivity;
            resourceInputs["maxNumResults"] = state?.maxNumResults;
            resourceInputs["metadata"] = state?.metadata;
            resourceInputs["modifiedAt"] = state?.modifiedAt;
            resourceInputs["modifiedBy"] = state?.modifiedBy;
            resourceInputs["namespace"] = state?.namespace;
            resourceInputs["paused"] = state?.paused;
            resourceInputs["publicEndpointId"] = state?.publicEndpointId;
            resourceInputs["publicEndpointParams"] = state?.publicEndpointParams;
            resourceInputs["reranking"] = state?.reranking;
            resourceInputs["rerankingModel"] = state?.rerankingModel;
            resourceInputs["retrievalOptions"] = state?.retrievalOptions;
            resourceInputs["rewriteModel"] = state?.rewriteModel;
            resourceInputs["rewriteQuery"] = state?.rewriteQuery;
            resourceInputs["scoreThreshold"] = state?.scoreThreshold;
            resourceInputs["source"] = state?.source;
            resourceInputs["sourceParams"] = state?.sourceParams;
            resourceInputs["status"] = state?.status;
            resourceInputs["summarization"] = state?.summarization;
            resourceInputs["summarizationModel"] = state?.summarizationModel;
            resourceInputs["syncInterval"] = state?.syncInterval;
            resourceInputs["systemPromptAisearch"] = state?.systemPromptAisearch;
            resourceInputs["systemPromptIndexSummarization"] = state?.systemPromptIndexSummarization;
            resourceInputs["systemPromptRewriteQuery"] = state?.systemPromptRewriteQuery;
            resourceInputs["tokenId"] = state?.tokenId;
            resourceInputs["type"] = state?.type;
            resourceInputs["vectorizeName"] = state?.vectorizeName;
        }
        else {
            const args = argsOrState;
            if (args?.accountId === undefined && !opts.urn) {
                throw new Error("Missing required property 'accountId'");
            }
            if (args?.aiSearchInstanceId === undefined && !opts.urn) {
                throw new Error("Missing required property 'aiSearchInstanceId'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["aiGatewayId"] = args?.aiGatewayId;
            resourceInputs["aiSearchInstanceId"] = args?.aiSearchInstanceId;
            resourceInputs["aisearchModel"] = args?.aisearchModel;
            resourceInputs["cache"] = args?.cache;
            resourceInputs["cacheThreshold"] = args?.cacheThreshold;
            resourceInputs["chunk"] = args?.chunk;
            resourceInputs["chunkOverlap"] = args?.chunkOverlap;
            resourceInputs["chunkSize"] = args?.chunkSize;
            resourceInputs["customMetadatas"] = args?.customMetadatas;
            resourceInputs["embeddingModel"] = args?.embeddingModel;
            resourceInputs["fusionMethod"] = args?.fusionMethod;
            resourceInputs["hybridSearchEnabled"] = args?.hybridSearchEnabled;
            resourceInputs["indexMethod"] = args?.indexMethod;
            resourceInputs["indexingOptions"] = args?.indexingOptions;
            resourceInputs["maxNumResults"] = args?.maxNumResults;
            resourceInputs["metadata"] = args?.metadata;
            resourceInputs["paused"] = args?.paused;
            resourceInputs["publicEndpointParams"] = args?.publicEndpointParams;
            resourceInputs["reranking"] = args?.reranking;
            resourceInputs["rerankingModel"] = args?.rerankingModel;
            resourceInputs["retrievalOptions"] = args?.retrievalOptions;
            resourceInputs["rewriteModel"] = args?.rewriteModel;
            resourceInputs["rewriteQuery"] = args?.rewriteQuery;
            resourceInputs["scoreThreshold"] = args?.scoreThreshold;
            resourceInputs["source"] = args?.source;
            resourceInputs["sourceParams"] = args?.sourceParams;
            resourceInputs["summarization"] = args?.summarization;
            resourceInputs["summarizationModel"] = args?.summarizationModel;
            resourceInputs["syncInterval"] = args?.syncInterval;
            resourceInputs["systemPromptAisearch"] = args?.systemPromptAisearch;
            resourceInputs["systemPromptIndexSummarization"] = args?.systemPromptIndexSummarization;
            resourceInputs["systemPromptRewriteQuery"] = args?.systemPromptRewriteQuery;
            resourceInputs["tokenId"] = args?.tokenId;
            resourceInputs["type"] = args?.type;
            resourceInputs["createdAt"] = undefined /*out*/;
            resourceInputs["createdBy"] = undefined /*out*/;
            resourceInputs["enable"] = undefined /*out*/;
            resourceInputs["engineVersion"] = undefined /*out*/;
            resourceInputs["lastActivity"] = undefined /*out*/;
            resourceInputs["modifiedAt"] = undefined /*out*/;
            resourceInputs["modifiedBy"] = undefined /*out*/;
            resourceInputs["namespace"] = undefined /*out*/;
            resourceInputs["publicEndpointId"] = undefined /*out*/;
            resourceInputs["status"] = undefined /*out*/;
            resourceInputs["vectorizeName"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(AiSearchInstance.__pulumiType, name, resourceInputs, opts);
    }
}
exports.AiSearchInstance = AiSearchInstance;
