// @pulumi/cloudflare/byoIpPrefix.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ByoIpPrefix = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `IP Prefixes: BGP On Demand Read`
 * - `IP Prefixes: BGP On Demand Write`
 * - `IP Prefixes: Read`
 * - `IP Prefixes: Write`
 * - `Magic Transit Read`
 * - `Magic Transit Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleByoIpPrefix = new cloudflare.ByoIpPrefix("example_byo_ip_prefix", {
 *     accountId: "258def64c72dae45f3e4c8516e2111f2",
 *     asn: 13335,
 *     cidr: "192.0.2.0/24",
 *     delegateLoaCreation: true,
 *     description: "Internal test prefix",
 *     loaDocumentId: "d933b1530bc56c9953cf8ce166da8004",
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/byoIpPrefix:ByoIpPrefix example '<account_id>/<prefix_id>'
 * ```
 */
class ByoIpPrefix extends pulumi.CustomResource {
    /**
     * Get an existing ByoIpPrefix resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ByoIpPrefix(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/byoIpPrefix:ByoIpPrefix';
    /**
     * Returns true if the given object is an instance of ByoIpPrefix.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ByoIpPrefix.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["advertised"] = state?.advertised;
            resourceInputs["advertisedModifiedAt"] = state?.advertisedModifiedAt;
            resourceInputs["approved"] = state?.approved;
            resourceInputs["asn"] = state?.asn;
            resourceInputs["cidr"] = state?.cidr;
            resourceInputs["createdAt"] = state?.createdAt;
            resourceInputs["delegateLoaCreation"] = state?.delegateLoaCreation;
            resourceInputs["description"] = state?.description;
            resourceInputs["irrValidationState"] = state?.irrValidationState;
            resourceInputs["loaDocumentId"] = state?.loaDocumentId;
            resourceInputs["modifiedAt"] = state?.modifiedAt;
            resourceInputs["onDemandEnabled"] = state?.onDemandEnabled;
            resourceInputs["onDemandLocked"] = state?.onDemandLocked;
            resourceInputs["ownershipValidationState"] = state?.ownershipValidationState;
            resourceInputs["ownershipValidationToken"] = state?.ownershipValidationToken;
            resourceInputs["rpkiValidationState"] = state?.rpkiValidationState;
        }
        else {
            const args = argsOrState;
            if (args?.asn === undefined && !opts.urn) {
                throw new Error("Missing required property 'asn'");
            }
            if (args?.cidr === undefined && !opts.urn) {
                throw new Error("Missing required property 'cidr'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["asn"] = args?.asn;
            resourceInputs["cidr"] = args?.cidr;
            resourceInputs["delegateLoaCreation"] = args?.delegateLoaCreation;
            resourceInputs["description"] = args?.description;
            resourceInputs["loaDocumentId"] = args?.loaDocumentId;
            resourceInputs["advertised"] = undefined /*out*/;
            resourceInputs["advertisedModifiedAt"] = undefined /*out*/;
            resourceInputs["approved"] = undefined /*out*/;
            resourceInputs["createdAt"] = undefined /*out*/;
            resourceInputs["irrValidationState"] = undefined /*out*/;
            resourceInputs["modifiedAt"] = undefined /*out*/;
            resourceInputs["onDemandEnabled"] = undefined /*out*/;
            resourceInputs["onDemandLocked"] = undefined /*out*/;
            resourceInputs["ownershipValidationState"] = undefined /*out*/;
            resourceInputs["ownershipValidationToken"] = undefined /*out*/;
            resourceInputs["rpkiValidationState"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(ByoIpPrefix.__pulumiType, name, resourceInputs, opts);
    }
}
exports.ByoIpPrefix = ByoIpPrefix;
