// @pulumi/cloudflare/cloudforceOneRequestPriority.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudforceOneRequestPriority = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Cloudforce One Read`
 * - `Cloudforce One Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleCloudforceOneRequestPriority = new cloudflare.CloudforceOneRequestPriority("example_cloudforce_one_request_priority", {
 *     accountId: "023e105f4ecef8ad9ca31a8372d0c353",
 *     labels: [
 *         "DoS",
 *         "CVE",
 *     ],
 *     priority: 1,
 *     requirement: "DoS attacks carried out by CVEs",
 *     tlp: "clear",
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/cloudforceOneRequestPriority:CloudforceOneRequestPriority example '<account_id>/<priority_id>'
 * ```
 */
class CloudforceOneRequestPriority extends pulumi.CustomResource {
    /**
     * Get an existing CloudforceOneRequestPriority resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new CloudforceOneRequestPriority(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/cloudforceOneRequestPriority:CloudforceOneRequestPriority';
    /**
     * Returns true if the given object is an instance of CloudforceOneRequestPriority.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === CloudforceOneRequestPriority.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["completed"] = state?.completed;
            resourceInputs["content"] = state?.content;
            resourceInputs["created"] = state?.created;
            resourceInputs["labels"] = state?.labels;
            resourceInputs["messageTokens"] = state?.messageTokens;
            resourceInputs["priority"] = state?.priority;
            resourceInputs["readableId"] = state?.readableId;
            resourceInputs["request"] = state?.request;
            resourceInputs["requirement"] = state?.requirement;
            resourceInputs["status"] = state?.status;
            resourceInputs["summary"] = state?.summary;
            resourceInputs["tlp"] = state?.tlp;
            resourceInputs["tokens"] = state?.tokens;
            resourceInputs["updated"] = state?.updated;
        }
        else {
            const args = argsOrState;
            if (args?.labels === undefined && !opts.urn) {
                throw new Error("Missing required property 'labels'");
            }
            if (args?.priority === undefined && !opts.urn) {
                throw new Error("Missing required property 'priority'");
            }
            if (args?.requirement === undefined && !opts.urn) {
                throw new Error("Missing required property 'requirement'");
            }
            if (args?.tlp === undefined && !opts.urn) {
                throw new Error("Missing required property 'tlp'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["labels"] = args?.labels;
            resourceInputs["priority"] = args?.priority;
            resourceInputs["requirement"] = args?.requirement;
            resourceInputs["tlp"] = args?.tlp;
            resourceInputs["completed"] = undefined /*out*/;
            resourceInputs["content"] = undefined /*out*/;
            resourceInputs["created"] = undefined /*out*/;
            resourceInputs["messageTokens"] = undefined /*out*/;
            resourceInputs["readableId"] = undefined /*out*/;
            resourceInputs["request"] = undefined /*out*/;
            resourceInputs["status"] = undefined /*out*/;
            resourceInputs["summary"] = undefined /*out*/;
            resourceInputs["tokens"] = undefined /*out*/;
            resourceInputs["updated"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(CloudforceOneRequestPriority.__pulumiType, name, resourceInputs, opts);
    }
}
exports.CloudforceOneRequestPriority = CloudforceOneRequestPriority;
