// @pulumi/cloudflare/cloudforceOneRequest.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudforceOneRequest = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Cloudforce One Read`
 * - `Cloudforce One Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleCloudforceOneRequest = new cloudflare.CloudforceOneRequest("example_cloudforce_one_request", {
 *     accountId: "023e105f4ecef8ad9ca31a8372d0c353",
 *     content: "What regions were most effected by the recent DoS?",
 *     priority: "routine",
 *     requestType: "Victomology",
 *     summary: "DoS attack",
 *     tlp: "clear",
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/cloudforceOneRequest:CloudforceOneRequest example '<account_id>/<request_id>'
 * ```
 */
class CloudforceOneRequest extends pulumi.CustomResource {
    /**
     * Get an existing CloudforceOneRequest resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new CloudforceOneRequest(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/cloudforceOneRequest:CloudforceOneRequest';
    /**
     * Returns true if the given object is an instance of CloudforceOneRequest.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === CloudforceOneRequest.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["completed"] = state?.completed;
            resourceInputs["content"] = state?.content;
            resourceInputs["created"] = state?.created;
            resourceInputs["messageTokens"] = state?.messageTokens;
            resourceInputs["priority"] = state?.priority;
            resourceInputs["readableId"] = state?.readableId;
            resourceInputs["request"] = state?.request;
            resourceInputs["requestType"] = state?.requestType;
            resourceInputs["status"] = state?.status;
            resourceInputs["summary"] = state?.summary;
            resourceInputs["tlp"] = state?.tlp;
            resourceInputs["tokens"] = state?.tokens;
            resourceInputs["updated"] = state?.updated;
        }
        else {
            const args = argsOrState;
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["content"] = args?.content;
            resourceInputs["priority"] = args?.priority;
            resourceInputs["requestType"] = args?.requestType;
            resourceInputs["summary"] = args?.summary;
            resourceInputs["tlp"] = args?.tlp;
            resourceInputs["completed"] = undefined /*out*/;
            resourceInputs["created"] = undefined /*out*/;
            resourceInputs["messageTokens"] = undefined /*out*/;
            resourceInputs["readableId"] = undefined /*out*/;
            resourceInputs["request"] = undefined /*out*/;
            resourceInputs["status"] = undefined /*out*/;
            resourceInputs["tokens"] = undefined /*out*/;
            resourceInputs["updated"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(CloudforceOneRequest.__pulumiType, name, resourceInputs, opts);
    }
}
exports.CloudforceOneRequest = CloudforceOneRequest;
