// @pulumi/cloudflare/connectivityDirectoryService.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectivityDirectoryService = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleConnectivityDirectoryService = new cloudflare.ConnectivityDirectoryService("example_connectivity_directory_service", {
 *     accountId: "023e105f4ecef8ad9ca31a8372d0c353",
 *     host: {
 *         ipv4: "10.0.0.1",
 *         network: {
 *             tunnelId: "0191dce4-9ab4-7fce-b660-8e5dec5172da",
 *         },
 *     },
 *     name: "web-app",
 *     type: "http",
 *     httpPort: 8080,
 *     httpsPort: 8443,
 *     tlsSettings: {
 *         certVerificationMode: "verify_full",
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/connectivityDirectoryService:ConnectivityDirectoryService example '<account_id>/<service_id>'
 * ```
 */
class ConnectivityDirectoryService extends pulumi.CustomResource {
    /**
     * Get an existing ConnectivityDirectoryService resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ConnectivityDirectoryService(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/connectivityDirectoryService:ConnectivityDirectoryService';
    /**
     * Returns true if the given object is an instance of ConnectivityDirectoryService.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ConnectivityDirectoryService.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["appProtocol"] = state?.appProtocol;
            resourceInputs["createdAt"] = state?.createdAt;
            resourceInputs["host"] = state?.host;
            resourceInputs["httpPort"] = state?.httpPort;
            resourceInputs["httpsPort"] = state?.httpsPort;
            resourceInputs["name"] = state?.name;
            resourceInputs["serviceId"] = state?.serviceId;
            resourceInputs["tcpPort"] = state?.tcpPort;
            resourceInputs["tlsSettings"] = state?.tlsSettings;
            resourceInputs["type"] = state?.type;
            resourceInputs["updatedAt"] = state?.updatedAt;
        }
        else {
            const args = argsOrState;
            if (args?.host === undefined && !opts.urn) {
                throw new Error("Missing required property 'host'");
            }
            if (args?.name === undefined && !opts.urn) {
                throw new Error("Missing required property 'name'");
            }
            if (args?.type === undefined && !opts.urn) {
                throw new Error("Missing required property 'type'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["appProtocol"] = args?.appProtocol;
            resourceInputs["host"] = args?.host;
            resourceInputs["httpPort"] = args?.httpPort;
            resourceInputs["httpsPort"] = args?.httpsPort;
            resourceInputs["name"] = args?.name;
            resourceInputs["tcpPort"] = args?.tcpPort;
            resourceInputs["tlsSettings"] = args?.tlsSettings;
            resourceInputs["type"] = args?.type;
            resourceInputs["createdAt"] = undefined /*out*/;
            resourceInputs["serviceId"] = undefined /*out*/;
            resourceInputs["updatedAt"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(ConnectivityDirectoryService.__pulumiType, name, resourceInputs, opts);
    }
}
exports.ConnectivityDirectoryService = ConnectivityDirectoryService;
