// @pulumi/cloudflare/d1database.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.D1Database = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `D1 Read`
 * - `D1 Write`
 *
 * !> When a D1 Database is replaced all the data is lost. Please ensure you have a
 *    backup of your data before replacing a D1 Database.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleD1Database = new cloudflare.D1Database("example_d1_database", {
 *     accountId: "023e105f4ecef8ad9ca31a8372d0c353",
 *     name: "my-database",
 *     jurisdiction: "eu",
 *     primaryLocationHint: "wnam",
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/d1Database:D1Database example '<account_id>/<database_id>'
 * ```
 */
class D1Database extends pulumi.CustomResource {
    /**
     * Get an existing D1Database resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new D1Database(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/d1Database:D1Database';
    /**
     * Returns true if the given object is an instance of D1Database.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === D1Database.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["createdAt"] = state?.createdAt;
            resourceInputs["fileSize"] = state?.fileSize;
            resourceInputs["jurisdiction"] = state?.jurisdiction;
            resourceInputs["name"] = state?.name;
            resourceInputs["numTables"] = state?.numTables;
            resourceInputs["primaryLocationHint"] = state?.primaryLocationHint;
            resourceInputs["readReplication"] = state?.readReplication;
            resourceInputs["uuid"] = state?.uuid;
            resourceInputs["version"] = state?.version;
        }
        else {
            const args = argsOrState;
            if (args?.name === undefined && !opts.urn) {
                throw new Error("Missing required property 'name'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["jurisdiction"] = args?.jurisdiction;
            resourceInputs["name"] = args?.name;
            resourceInputs["primaryLocationHint"] = args?.primaryLocationHint;
            resourceInputs["readReplication"] = args?.readReplication;
            resourceInputs["createdAt"] = undefined /*out*/;
            resourceInputs["fileSize"] = undefined /*out*/;
            resourceInputs["numTables"] = undefined /*out*/;
            resourceInputs["uuid"] = undefined /*out*/;
            resourceInputs["version"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(D1Database.__pulumiType, name, resourceInputs, opts);
    }
}
exports.D1Database = D1Database;
