// @pulumi/cloudflare/dnsZoneTransfersIncoming.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DnsZoneTransfersIncoming = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `DNS Read`
 * - `DNS Write`
 * - `Zone Settings Read`
 * - `Zone Settings Write`
 * - `Zone Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleDnsZoneTransfersIncoming = new cloudflare.DnsZoneTransfersIncoming("example_dns_zone_transfers_incoming", {
 *     zoneId: "269d8f4853475ca241c4e730be286b20",
 *     autoRefreshSeconds: 86400,
 *     name: "www.example.com.",
 *     peers: [
 *         "23ff594956f20c2a721606e94745a8aa",
 *         "00920f38ce07c2e2f4df50b1f61d4194",
 *     ],
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/dnsZoneTransfersIncoming:DnsZoneTransfersIncoming example '<zone_id>'
 * ```
 */
class DnsZoneTransfersIncoming extends pulumi.CustomResource {
    /**
     * Get an existing DnsZoneTransfersIncoming resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new DnsZoneTransfersIncoming(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/dnsZoneTransfersIncoming:DnsZoneTransfersIncoming';
    /**
     * Returns true if the given object is an instance of DnsZoneTransfersIncoming.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === DnsZoneTransfersIncoming.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["autoRefreshSeconds"] = state?.autoRefreshSeconds;
            resourceInputs["checkedTime"] = state?.checkedTime;
            resourceInputs["createdTime"] = state?.createdTime;
            resourceInputs["modifiedTime"] = state?.modifiedTime;
            resourceInputs["name"] = state?.name;
            resourceInputs["peers"] = state?.peers;
            resourceInputs["soaSerial"] = state?.soaSerial;
            resourceInputs["zoneId"] = state?.zoneId;
        }
        else {
            const args = argsOrState;
            if (args?.name === undefined && !opts.urn) {
                throw new Error("Missing required property 'name'");
            }
            if (args?.peers === undefined && !opts.urn) {
                throw new Error("Missing required property 'peers'");
            }
            resourceInputs["autoRefreshSeconds"] = args?.autoRefreshSeconds;
            resourceInputs["name"] = args?.name;
            resourceInputs["peers"] = args?.peers;
            resourceInputs["zoneId"] = args?.zoneId;
            resourceInputs["checkedTime"] = undefined /*out*/;
            resourceInputs["createdTime"] = undefined /*out*/;
            resourceInputs["modifiedTime"] = undefined /*out*/;
            resourceInputs["soaSerial"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(DnsZoneTransfersIncoming.__pulumiType, name, resourceInputs, opts);
    }
}
exports.DnsZoneTransfersIncoming = DnsZoneTransfersIncoming;
