// @pulumi/cloudflare/emailSecurityBlockSender.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmailSecurityBlockSender = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Cloud Email Security: Read`
 * - `Cloud Email Security: Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleEmailSecurityBlockSender = new cloudflare.EmailSecurityBlockSender("example_email_security_block_sender", {
 *     accountId: "023e105f4ecef8ad9ca31a8372d0c353",
 *     isRegex: false,
 *     pattern: "test@example.com",
 *     patternType: "EMAIL",
 *     comments: "block sender with email test@example.com",
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/emailSecurityBlockSender:EmailSecurityBlockSender example '<account_id>/<pattern_id>'
 * ```
 */
class EmailSecurityBlockSender extends pulumi.CustomResource {
    /**
     * Get an existing EmailSecurityBlockSender resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new EmailSecurityBlockSender(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/emailSecurityBlockSender:EmailSecurityBlockSender';
    /**
     * Returns true if the given object is an instance of EmailSecurityBlockSender.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === EmailSecurityBlockSender.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["comments"] = state?.comments;
            resourceInputs["createdAt"] = state?.createdAt;
            resourceInputs["isRegex"] = state?.isRegex;
            resourceInputs["lastModified"] = state?.lastModified;
            resourceInputs["pattern"] = state?.pattern;
            resourceInputs["patternType"] = state?.patternType;
        }
        else {
            const args = argsOrState;
            if (args?.isRegex === undefined && !opts.urn) {
                throw new Error("Missing required property 'isRegex'");
            }
            if (args?.pattern === undefined && !opts.urn) {
                throw new Error("Missing required property 'pattern'");
            }
            if (args?.patternType === undefined && !opts.urn) {
                throw new Error("Missing required property 'patternType'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["comments"] = args?.comments;
            resourceInputs["isRegex"] = args?.isRegex;
            resourceInputs["pattern"] = args?.pattern;
            resourceInputs["patternType"] = args?.patternType;
            resourceInputs["createdAt"] = undefined /*out*/;
            resourceInputs["lastModified"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(EmailSecurityBlockSender.__pulumiType, name, resourceInputs, opts);
    }
}
exports.EmailSecurityBlockSender = EmailSecurityBlockSender;
