// @pulumi/cloudflare/loadBalancer.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoadBalancer = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Load Balancers Read`
 * - `Load Balancers Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleLoadBalancer = new cloudflare.LoadBalancer("example_load_balancer", {
 *     zoneId: "699d98642c564d2e855e9661899b7252",
 *     defaultPools: [
 *         "17b5962d775c646f3f9725cbc7a53df4",
 *         "9290f38c5d07c2e2f4df57b1f61d4196",
 *         "00920f38ce07c2e2f4df50b1f61d4194",
 *     ],
 *     fallbackPool: "fallback_pool",
 *     name: "www.example.com",
 *     adaptiveRouting: {
 *         failoverAcrossPools: true,
 *     },
 *     countryPools: {
 *         GB: ["abd90f38ced07c2e2f4df50b1f61d4194"],
 *         US: [
 *             "de90f38ced07c2e2f4df50b1f61d4194",
 *             "00920f38ce07c2e2f4df50b1f61d4194",
 *         ],
 *     },
 *     description: "Load Balancer for www.example.com",
 *     locationStrategy: {
 *         mode: "resolver_ip",
 *         preferEcs: "always",
 *     },
 *     networks: ["string"],
 *     popPools: {
 *         LAX: [
 *             "de90f38ced07c2e2f4df50b1f61d4194",
 *             "9290f38c5d07c2e2f4df57b1f61d4196",
 *         ],
 *         LHR: [
 *             "abd90f38ced07c2e2f4df50b1f61d4194",
 *             "f9138c5d07c2e2f4df57b1f61d4196",
 *         ],
 *         SJC: ["00920f38ce07c2e2f4df50b1f61d4194"],
 *     },
 *     proxied: true,
 *     randomSteering: {
 *         defaultWeight: 0.2,
 *         poolWeights: {
 *             "9290f38c5d07c2e2f4df57b1f61d4196": 0.5,
 *             de90f38ced07c2e2f4df50b1f61d4194: 0.3,
 *         },
 *     },
 *     regionPools: {
 *         ENAM: ["00920f38ce07c2e2f4df50b1f61d4194"],
 *         WNAM: [
 *             "de90f38ced07c2e2f4df50b1f61d4194",
 *             "9290f38c5d07c2e2f4df57b1f61d4196",
 *         ],
 *     },
 *     rules: [{
 *         condition: "http.request.uri.path contains \"/testing\"",
 *         disabled: true,
 *         fixedResponse: {
 *             contentType: "application/json",
 *             location: "www.example.com",
 *             messageBody: "Testing Hello",
 *             statusCode: 0,
 *         },
 *         name: "route the path /testing to testing datacenter.",
 *         overrides: {
 *             adaptiveRouting: {
 *                 failoverAcrossPools: true,
 *             },
 *             countryPools: {
 *                 GB: ["abd90f38ced07c2e2f4df50b1f61d4194"],
 *                 US: [
 *                     "de90f38ced07c2e2f4df50b1f61d4194",
 *                     "00920f38ce07c2e2f4df50b1f61d4194",
 *                 ],
 *             },
 *             defaultPools: [
 *                 "17b5962d775c646f3f9725cbc7a53df4",
 *                 "9290f38c5d07c2e2f4df57b1f61d4196",
 *                 "00920f38ce07c2e2f4df50b1f61d4194",
 *             ],
 *             fallbackPool: "fallback_pool",
 *             locationStrategy: {
 *                 mode: "resolver_ip",
 *                 preferEcs: "always",
 *             },
 *             popPools: {
 *                 LAX: [
 *                     "de90f38ced07c2e2f4df50b1f61d4194",
 *                     "9290f38c5d07c2e2f4df57b1f61d4196",
 *                 ],
 *                 LHR: [
 *                     "abd90f38ced07c2e2f4df50b1f61d4194",
 *                     "f9138c5d07c2e2f4df57b1f61d4196",
 *                 ],
 *                 SJC: ["00920f38ce07c2e2f4df50b1f61d4194"],
 *             },
 *             randomSteering: {
 *                 defaultWeight: 0.2,
 *                 poolWeights: {
 *                     "9290f38c5d07c2e2f4df57b1f61d4196": 0.5,
 *                     de90f38ced07c2e2f4df50b1f61d4194: 0.3,
 *                 },
 *             },
 *             regionPools: {
 *                 ENAM: ["00920f38ce07c2e2f4df50b1f61d4194"],
 *                 WNAM: [
 *                     "de90f38ced07c2e2f4df50b1f61d4194",
 *                     "9290f38c5d07c2e2f4df57b1f61d4196",
 *                 ],
 *             },
 *             sessionAffinity: "cookie",
 *             sessionAffinityAttributes: {
 *                 drainDuration: 100,
 *                 headers: ["x"],
 *                 requireAllHeaders: true,
 *                 samesite: "Auto",
 *                 secure: "Auto",
 *                 zeroDowntimeFailover: "sticky",
 *             },
 *             sessionAffinityTtl: 1800,
 *             steeringPolicy: "dynamic_latency",
 *             ttl: 30,
 *         },
 *         priority: 0,
 *         terminates: true,
 *     }],
 *     sessionAffinity: "cookie",
 *     sessionAffinityAttributes: {
 *         drainDuration: 100,
 *         headers: ["x"],
 *         requireAllHeaders: true,
 *         samesite: "Auto",
 *         secure: "Auto",
 *         zeroDowntimeFailover: "sticky",
 *     },
 *     sessionAffinityTtl: 1800,
 *     steeringPolicy: "dynamic_latency",
 *     ttl: 30,
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/loadBalancer:LoadBalancer example '<zone_id>/<load_balancer_id>'
 * ```
 */
class LoadBalancer extends pulumi.CustomResource {
    /**
     * Get an existing LoadBalancer resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new LoadBalancer(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/loadBalancer:LoadBalancer';
    /**
     * Returns true if the given object is an instance of LoadBalancer.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === LoadBalancer.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["adaptiveRouting"] = state?.adaptiveRouting;
            resourceInputs["countryPools"] = state?.countryPools;
            resourceInputs["createdOn"] = state?.createdOn;
            resourceInputs["defaultPools"] = state?.defaultPools;
            resourceInputs["description"] = state?.description;
            resourceInputs["enabled"] = state?.enabled;
            resourceInputs["fallbackPool"] = state?.fallbackPool;
            resourceInputs["locationStrategy"] = state?.locationStrategy;
            resourceInputs["modifiedOn"] = state?.modifiedOn;
            resourceInputs["name"] = state?.name;
            resourceInputs["networks"] = state?.networks;
            resourceInputs["popPools"] = state?.popPools;
            resourceInputs["proxied"] = state?.proxied;
            resourceInputs["randomSteering"] = state?.randomSteering;
            resourceInputs["regionPools"] = state?.regionPools;
            resourceInputs["rules"] = state?.rules;
            resourceInputs["sessionAffinity"] = state?.sessionAffinity;
            resourceInputs["sessionAffinityAttributes"] = state?.sessionAffinityAttributes;
            resourceInputs["sessionAffinityTtl"] = state?.sessionAffinityTtl;
            resourceInputs["steeringPolicy"] = state?.steeringPolicy;
            resourceInputs["ttl"] = state?.ttl;
            resourceInputs["zoneId"] = state?.zoneId;
            resourceInputs["zoneName"] = state?.zoneName;
        }
        else {
            const args = argsOrState;
            if (args?.defaultPools === undefined && !opts.urn) {
                throw new Error("Missing required property 'defaultPools'");
            }
            if (args?.fallbackPool === undefined && !opts.urn) {
                throw new Error("Missing required property 'fallbackPool'");
            }
            if (args?.name === undefined && !opts.urn) {
                throw new Error("Missing required property 'name'");
            }
            resourceInputs["adaptiveRouting"] = args?.adaptiveRouting;
            resourceInputs["countryPools"] = args?.countryPools;
            resourceInputs["defaultPools"] = args?.defaultPools;
            resourceInputs["description"] = args?.description;
            resourceInputs["enabled"] = args?.enabled;
            resourceInputs["fallbackPool"] = args?.fallbackPool;
            resourceInputs["locationStrategy"] = args?.locationStrategy;
            resourceInputs["name"] = args?.name;
            resourceInputs["networks"] = args?.networks;
            resourceInputs["popPools"] = args?.popPools;
            resourceInputs["proxied"] = args?.proxied;
            resourceInputs["randomSteering"] = args?.randomSteering;
            resourceInputs["regionPools"] = args?.regionPools;
            resourceInputs["rules"] = args?.rules;
            resourceInputs["sessionAffinity"] = args?.sessionAffinity;
            resourceInputs["sessionAffinityAttributes"] = args?.sessionAffinityAttributes;
            resourceInputs["sessionAffinityTtl"] = args?.sessionAffinityTtl;
            resourceInputs["steeringPolicy"] = args?.steeringPolicy;
            resourceInputs["ttl"] = args?.ttl;
            resourceInputs["zoneId"] = args?.zoneId;
            resourceInputs["createdOn"] = undefined /*out*/;
            resourceInputs["modifiedOn"] = undefined /*out*/;
            resourceInputs["zoneName"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(LoadBalancer.__pulumiType, name, resourceInputs, opts);
    }
}
exports.LoadBalancer = LoadBalancer;
