// @pulumi/cloudflare/loadBalancerMonitor.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoadBalancerMonitor = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Load Balancing: Monitors and Pools Read`
 * - `Load Balancing: Monitors and Pools Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleLoadBalancerMonitor = new cloudflare.LoadBalancerMonitor("example_load_balancer_monitor", {
 *     accountId: "023e105f4ecef8ad9ca31a8372d0c353",
 *     allowInsecure: true,
 *     consecutiveDown: 0,
 *     consecutiveUp: 0,
 *     description: "Login page monitor",
 *     expectedBody: "alive",
 *     expectedCodes: "2xx",
 *     followRedirects: true,
 *     header: {
 *         Host: ["example.com"],
 *         "X-App-ID": ["abc123"],
 *     },
 *     interval: 0,
 *     method: "GET",
 *     path: "/health",
 *     port: 0,
 *     probeZone: "example.com",
 *     retries: 0,
 *     timeout: 0,
 *     type: "https",
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/loadBalancerMonitor:LoadBalancerMonitor example '<account_id>/<monitor_id>'
 * ```
 */
class LoadBalancerMonitor extends pulumi.CustomResource {
    /**
     * Get an existing LoadBalancerMonitor resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new LoadBalancerMonitor(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/loadBalancerMonitor:LoadBalancerMonitor';
    /**
     * Returns true if the given object is an instance of LoadBalancerMonitor.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === LoadBalancerMonitor.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["allowInsecure"] = state?.allowInsecure;
            resourceInputs["consecutiveDown"] = state?.consecutiveDown;
            resourceInputs["consecutiveUp"] = state?.consecutiveUp;
            resourceInputs["createdOn"] = state?.createdOn;
            resourceInputs["description"] = state?.description;
            resourceInputs["expectedBody"] = state?.expectedBody;
            resourceInputs["expectedCodes"] = state?.expectedCodes;
            resourceInputs["followRedirects"] = state?.followRedirects;
            resourceInputs["header"] = state?.header;
            resourceInputs["interval"] = state?.interval;
            resourceInputs["method"] = state?.method;
            resourceInputs["modifiedOn"] = state?.modifiedOn;
            resourceInputs["path"] = state?.path;
            resourceInputs["port"] = state?.port;
            resourceInputs["probeZone"] = state?.probeZone;
            resourceInputs["retries"] = state?.retries;
            resourceInputs["timeout"] = state?.timeout;
            resourceInputs["type"] = state?.type;
        }
        else {
            const args = argsOrState;
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["allowInsecure"] = args?.allowInsecure;
            resourceInputs["consecutiveDown"] = args?.consecutiveDown;
            resourceInputs["consecutiveUp"] = args?.consecutiveUp;
            resourceInputs["description"] = args?.description;
            resourceInputs["expectedBody"] = args?.expectedBody;
            resourceInputs["expectedCodes"] = args?.expectedCodes;
            resourceInputs["followRedirects"] = args?.followRedirects;
            resourceInputs["header"] = args?.header;
            resourceInputs["interval"] = args?.interval;
            resourceInputs["method"] = args?.method;
            resourceInputs["path"] = args?.path;
            resourceInputs["port"] = args?.port;
            resourceInputs["probeZone"] = args?.probeZone;
            resourceInputs["retries"] = args?.retries;
            resourceInputs["timeout"] = args?.timeout;
            resourceInputs["type"] = args?.type;
            resourceInputs["createdOn"] = undefined /*out*/;
            resourceInputs["modifiedOn"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(LoadBalancerMonitor.__pulumiType, name, resourceInputs, opts);
    }
}
exports.LoadBalancerMonitor = LoadBalancerMonitor;
