// @pulumi/cloudflare/magicTransitSiteWan.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MagicTransitSiteWan = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Magic Transit Read`
 * - `Magic Transit Write`
 * - `Magic WAN Read`
 * - `Magic WAN Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleMagicTransitSiteWan = new cloudflare.MagicTransitSiteWan("example_magic_transit_site_wan", {
 *     accountId: "023e105f4ecef8ad9ca31a8372d0c353",
 *     siteId: "023e105f4ecef8ad9ca31a8372d0c353",
 *     physport: 1,
 *     name: "name",
 *     priority: 0,
 *     staticAddressing: {
 *         address: "192.0.2.0/24",
 *         gatewayAddress: "192.0.2.1",
 *         secondaryAddress: "192.0.2.0/24",
 *     },
 *     vlanTag: 42,
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/magicTransitSiteWan:MagicTransitSiteWan example '<account_id>/<site_id>/<wan_id>'
 * ```
 */
class MagicTransitSiteWan extends pulumi.CustomResource {
    /**
     * Get an existing MagicTransitSiteWan resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new MagicTransitSiteWan(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/magicTransitSiteWan:MagicTransitSiteWan';
    /**
     * Returns true if the given object is an instance of MagicTransitSiteWan.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === MagicTransitSiteWan.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["healthCheckRate"] = state?.healthCheckRate;
            resourceInputs["name"] = state?.name;
            resourceInputs["physport"] = state?.physport;
            resourceInputs["priority"] = state?.priority;
            resourceInputs["siteId"] = state?.siteId;
            resourceInputs["staticAddressing"] = state?.staticAddressing;
            resourceInputs["vlanTag"] = state?.vlanTag;
        }
        else {
            const args = argsOrState;
            if (args?.physport === undefined && !opts.urn) {
                throw new Error("Missing required property 'physport'");
            }
            if (args?.siteId === undefined && !opts.urn) {
                throw new Error("Missing required property 'siteId'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["name"] = args?.name;
            resourceInputs["physport"] = args?.physport;
            resourceInputs["priority"] = args?.priority;
            resourceInputs["siteId"] = args?.siteId;
            resourceInputs["staticAddressing"] = args?.staticAddressing;
            resourceInputs["vlanTag"] = args?.vlanTag;
            resourceInputs["healthCheckRate"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(MagicTransitSiteWan.__pulumiType, name, resourceInputs, opts);
    }
}
exports.MagicTransitSiteWan = MagicTransitSiteWan;
