// @pulumi/cloudflare/magicWanIpsecTunnel.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MagicWanIpsecTunnel = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleMagicWanIpsecTunnel = new cloudflare.MagicWanIpsecTunnel("example_magic_wan_ipsec_tunnel", {
 *     accountId: "023e105f4ecef8ad9ca31a8372d0c353",
 *     cloudflareEndpoint: "203.0.113.1",
 *     interfaceAddress: "192.0.2.0/31",
 *     name: "IPsec_1",
 *     automaticReturnRouting: true,
 *     bgp: {
 *         customerAsn: 0,
 *         extraPrefixes: ["string"],
 *         md5Key: "md5_key",
 *     },
 *     customRemoteIdentities: {
 *         fqdnId: "fqdn_id",
 *     },
 *     customerEndpoint: "203.0.113.1",
 *     description: "Tunnel for ISP X",
 *     healthCheck: {
 *         direction: "bidirectional",
 *         enabled: true,
 *         rate: "low",
 *         target: {
 *             saved: "203.0.113.1",
 *         },
 *         type: "request",
 *     },
 *     interfaceAddress6: "2606:54c1:7:0:a9fe:12d2:1:200/127",
 *     psk: "O3bwKSjnaoCxDoUxjcq4Rk8ZKkezQUiy",
 *     replayProtection: false,
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/magicWanIpsecTunnel:MagicWanIpsecTunnel example '<account_id>/<ipsec_tunnel_id>'
 * ```
 */
class MagicWanIpsecTunnel extends pulumi.CustomResource {
    /**
     * Get an existing MagicWanIpsecTunnel resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new MagicWanIpsecTunnel(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/magicWanIpsecTunnel:MagicWanIpsecTunnel';
    /**
     * Returns true if the given object is an instance of MagicWanIpsecTunnel.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === MagicWanIpsecTunnel.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["allowNullCipher"] = state?.allowNullCipher;
            resourceInputs["automaticReturnRouting"] = state?.automaticReturnRouting;
            resourceInputs["bgp"] = state?.bgp;
            resourceInputs["bgpStatus"] = state?.bgpStatus;
            resourceInputs["cloudflareEndpoint"] = state?.cloudflareEndpoint;
            resourceInputs["createdOn"] = state?.createdOn;
            resourceInputs["customRemoteIdentities"] = state?.customRemoteIdentities;
            resourceInputs["customerEndpoint"] = state?.customerEndpoint;
            resourceInputs["description"] = state?.description;
            resourceInputs["healthCheck"] = state?.healthCheck;
            resourceInputs["interfaceAddress"] = state?.interfaceAddress;
            resourceInputs["interfaceAddress6"] = state?.interfaceAddress6;
            resourceInputs["modifiedOn"] = state?.modifiedOn;
            resourceInputs["name"] = state?.name;
            resourceInputs["psk"] = state?.psk;
            resourceInputs["pskMetadata"] = state?.pskMetadata;
            resourceInputs["replayProtection"] = state?.replayProtection;
        }
        else {
            const args = argsOrState;
            if (args?.accountId === undefined && !opts.urn) {
                throw new Error("Missing required property 'accountId'");
            }
            if (args?.cloudflareEndpoint === undefined && !opts.urn) {
                throw new Error("Missing required property 'cloudflareEndpoint'");
            }
            if (args?.interfaceAddress === undefined && !opts.urn) {
                throw new Error("Missing required property 'interfaceAddress'");
            }
            if (args?.name === undefined && !opts.urn) {
                throw new Error("Missing required property 'name'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["automaticReturnRouting"] = args?.automaticReturnRouting;
            resourceInputs["bgp"] = args?.bgp;
            resourceInputs["cloudflareEndpoint"] = args?.cloudflareEndpoint;
            resourceInputs["customRemoteIdentities"] = args?.customRemoteIdentities;
            resourceInputs["customerEndpoint"] = args?.customerEndpoint;
            resourceInputs["description"] = args?.description;
            resourceInputs["healthCheck"] = args?.healthCheck;
            resourceInputs["interfaceAddress"] = args?.interfaceAddress;
            resourceInputs["interfaceAddress6"] = args?.interfaceAddress6;
            resourceInputs["name"] = args?.name;
            resourceInputs["psk"] = args?.psk ? pulumi.secret(args.psk) : undefined;
            resourceInputs["replayProtection"] = args?.replayProtection;
            resourceInputs["allowNullCipher"] = undefined /*out*/;
            resourceInputs["bgpStatus"] = undefined /*out*/;
            resourceInputs["createdOn"] = undefined /*out*/;
            resourceInputs["modifiedOn"] = undefined /*out*/;
            resourceInputs["pskMetadata"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const aliasOpts = { aliases: [{ type: "cloudflare:index/ipsecTunnel:IpsecTunnel" }] };
        opts = pulumi.mergeOptions(opts, aliasOpts);
        const secretOpts = { additionalSecretOutputs: ["psk"] };
        opts = pulumi.mergeOptions(opts, secretOpts);
        super(MagicWanIpsecTunnel.__pulumiType, name, resourceInputs, opts);
    }
}
exports.MagicWanIpsecTunnel = MagicWanIpsecTunnel;
