// @pulumi/cloudflare/notificationPolicy.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotificationPolicy = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Account Settings Read`
 * - `Account Settings Write`
 * - `Notifications Read`
 * - `Notifications Write`
 * - `Zero Trust: PII Read`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleNotificationPolicy = new cloudflare.NotificationPolicy("example_notification_policy", {
 *     accountId: "023e105f4ecef8ad9ca31a8372d0c353",
 *     alertType: "universal_ssl_event_type",
 *     enabled: true,
 *     mechanisms: {
 *         emails: [{
 *             id: "id",
 *         }],
 *         pagerduties: [{
 *             id: "f174e90afafe4643bbbc4a0ed4fc8415",
 *         }],
 *         webhooks: [{
 *             id: "f174e90afafe4643bbbc4a0ed4fc8415",
 *         }],
 *     },
 *     name: "SSL Notification Event Policy",
 *     alertInterval: "30m",
 *     description: "Something describing the policy.",
 *     filters: {
 *         actions: ["string"],
 *         affectedAsns: ["string"],
 *         affectedComponents: ["string"],
 *         affectedLocations: ["string"],
 *         airportCodes: ["string"],
 *         alertTriggerPreferences: ["string"],
 *         alertTriggerPreferencesValues: ["string"],
 *         enableds: ["string"],
 *         environments: ["string"],
 *         events: ["string"],
 *         eventSources: ["string"],
 *         eventTypes: ["string"],
 *         groupBies: ["string"],
 *         healthCheckIds: ["string"],
 *         incidentImpacts: ["INCIDENT_IMPACT_NONE"],
 *         inputIds: ["string"],
 *         insightClasses: ["string"],
 *         limits: ["string"],
 *         logoTags: ["string"],
 *         megabitsPerSeconds: ["string"],
 *         newHealths: ["string"],
 *         newStatuses: ["string"],
 *         packetsPerSeconds: ["string"],
 *         poolIds: ["string"],
 *         popNames: ["string"],
 *         products: ["string"],
 *         projectIds: ["string"],
 *         protocols: ["string"],
 *         queryTags: ["string"],
 *         requestsPerSeconds: ["string"],
 *         selectors: ["string"],
 *         services: ["string"],
 *         slos: ["99.9"],
 *         statuses: ["string"],
 *         targetHostnames: ["string"],
 *         targetIps: ["string"],
 *         targetZoneNames: ["string"],
 *         trafficExclusions: ["security_events"],
 *         tunnelIds: ["string"],
 *         tunnelNames: ["string"],
 *         types: ["string"],
 *         wheres: ["string"],
 *         zones: ["string"],
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/notificationPolicy:NotificationPolicy example '<account_id>/<policy_id>'
 * ```
 */
class NotificationPolicy extends pulumi.CustomResource {
    /**
     * Get an existing NotificationPolicy resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new NotificationPolicy(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/notificationPolicy:NotificationPolicy';
    /**
     * Returns true if the given object is an instance of NotificationPolicy.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === NotificationPolicy.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["alertInterval"] = state?.alertInterval;
            resourceInputs["alertType"] = state?.alertType;
            resourceInputs["created"] = state?.created;
            resourceInputs["description"] = state?.description;
            resourceInputs["enabled"] = state?.enabled;
            resourceInputs["filters"] = state?.filters;
            resourceInputs["mechanisms"] = state?.mechanisms;
            resourceInputs["modified"] = state?.modified;
            resourceInputs["name"] = state?.name;
        }
        else {
            const args = argsOrState;
            if (args?.alertType === undefined && !opts.urn) {
                throw new Error("Missing required property 'alertType'");
            }
            if (args?.mechanisms === undefined && !opts.urn) {
                throw new Error("Missing required property 'mechanisms'");
            }
            if (args?.name === undefined && !opts.urn) {
                throw new Error("Missing required property 'name'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["alertInterval"] = args?.alertInterval;
            resourceInputs["alertType"] = args?.alertType;
            resourceInputs["description"] = args?.description;
            resourceInputs["enabled"] = args?.enabled;
            resourceInputs["filters"] = args?.filters;
            resourceInputs["mechanisms"] = args?.mechanisms;
            resourceInputs["name"] = args?.name;
            resourceInputs["created"] = undefined /*out*/;
            resourceInputs["modified"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(NotificationPolicy.__pulumiType, name, resourceInputs, opts);
    }
}
exports.NotificationPolicy = NotificationPolicy;
