// @pulumi/cloudflare/pageRule.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PageRule = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Access: Apps and Policies Read`
 * - `Access: Apps and Policies Revoke`
 * - `Access: Apps and Policies Write`
 * - `Access: Mutual TLS Certificates Write`
 * - `Access: Organizations, Identity Providers, and Groups Write`
 * - `Analytics Read`
 * - `Apps Write`
 * - `Cache Purge`
 * - `DNS Read`
 * - `DNS Write`
 * - `Firewall Services Read`
 * - `Firewall Services Write`
 * - `Load Balancers Read`
 * - `Load Balancers Write`
 * - `Logs Read`
 * - `Logs Write`
 * - `Page Rules Read`
 * - `Page Rules Write`
 * - `SSL and Certificates Read`
 * - `SSL and Certificates Write`
 * - `Stream Read`
 * - `Stream Write`
 * - `Trust and Safety Read`
 * - `Trust and Safety Write`
 * - `Workers Routes Read`
 * - `Workers Routes Write`
 * - `Workers Scripts Read`
 * - `Workers Scripts Write`
 * - `Zaraz Admin`
 * - `Zaraz Edit`
 * - `Zaraz Read`
 * - `Zero Trust: PII Read`
 * - `Zone Read`
 * - `Zone Settings Read`
 * - `Zone Settings Write`
 * - `Zone Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const examplePageRule = new cloudflare.PageRule("example_page_rule", {
 *     zoneId: "023e105f4ecef8ad9ca31a8372d0c353",
 *     target: "example.com/*",
 *     priority: 1,
 *     status: "active",
 *     actions: {
 *         forwardingUrl: {
 *             url: "https://example.com/foo",
 *             statusCode: 301,
 *         },
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/pageRule:PageRule example '<zone_id>/<pagerule_id>'
 * ```
 */
class PageRule extends pulumi.CustomResource {
    /**
     * Get an existing PageRule resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new PageRule(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/pageRule:PageRule';
    /**
     * Returns true if the given object is an instance of PageRule.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === PageRule.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["actions"] = state?.actions;
            resourceInputs["createdOn"] = state?.createdOn;
            resourceInputs["modifiedOn"] = state?.modifiedOn;
            resourceInputs["priority"] = state?.priority;
            resourceInputs["status"] = state?.status;
            resourceInputs["target"] = state?.target;
            resourceInputs["zoneId"] = state?.zoneId;
        }
        else {
            const args = argsOrState;
            if (args?.actions === undefined && !opts.urn) {
                throw new Error("Missing required property 'actions'");
            }
            if (args?.target === undefined && !opts.urn) {
                throw new Error("Missing required property 'target'");
            }
            resourceInputs["actions"] = args?.actions;
            resourceInputs["priority"] = args?.priority;
            resourceInputs["status"] = args?.status;
            resourceInputs["target"] = args?.target;
            resourceInputs["zoneId"] = args?.zoneId;
            resourceInputs["createdOn"] = undefined /*out*/;
            resourceInputs["modifiedOn"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(PageRule.__pulumiType, name, resourceInputs, opts);
    }
}
exports.PageRule = PageRule;
