// @pulumi/cloudflare/pipelineStream.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PipelineStream = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Pipelines Read`
 * - `Pipelines Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const examplePipelineStream = new cloudflare.PipelineStream("example_pipeline_stream", {
 *     accountId: "0123105f4ecef8ad9ca31a8372d0c353",
 *     name: "my_stream",
 *     format: {
 *         type: "json",
 *         decimalEncoding: "number",
 *         timestampFormat: "rfc3339",
 *         unstructured: true,
 *     },
 *     http: {
 *         authentication: false,
 *         enabled: true,
 *         cors: {
 *             origins: ["string"],
 *         },
 *     },
 *     schema: {
 *         fields: [{
 *             type: "int32",
 *             metadataKey: "metadata_key",
 *             name: "name",
 *             required: true,
 *             sqlName: "sql_name",
 *         }],
 *         format: {
 *             type: "json",
 *             decimalEncoding: "number",
 *             timestampFormat: "rfc3339",
 *             unstructured: true,
 *         },
 *         inferred: true,
 *     },
 *     workerBinding: {
 *         enabled: true,
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/pipelineStream:PipelineStream example '<account_id>/<stream_id>'
 * ```
 */
class PipelineStream extends pulumi.CustomResource {
    /**
     * Get an existing PipelineStream resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new PipelineStream(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/pipelineStream:PipelineStream';
    /**
     * Returns true if the given object is an instance of PipelineStream.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === PipelineStream.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["createdAt"] = state?.createdAt;
            resourceInputs["endpoint"] = state?.endpoint;
            resourceInputs["format"] = state?.format;
            resourceInputs["http"] = state?.http;
            resourceInputs["modifiedAt"] = state?.modifiedAt;
            resourceInputs["name"] = state?.name;
            resourceInputs["schema"] = state?.schema;
            resourceInputs["version"] = state?.version;
            resourceInputs["workerBinding"] = state?.workerBinding;
        }
        else {
            const args = argsOrState;
            if (args?.name === undefined && !opts.urn) {
                throw new Error("Missing required property 'name'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["format"] = args?.format;
            resourceInputs["http"] = args?.http;
            resourceInputs["name"] = args?.name;
            resourceInputs["schema"] = args?.schema;
            resourceInputs["workerBinding"] = args?.workerBinding;
            resourceInputs["createdAt"] = undefined /*out*/;
            resourceInputs["endpoint"] = undefined /*out*/;
            resourceInputs["modifiedAt"] = undefined /*out*/;
            resourceInputs["version"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(PipelineStream.__pulumiType, name, resourceInputs, opts);
    }
}
exports.PipelineStream = PipelineStream;
