// @pulumi/cloudflare/r2bucketEventNotification.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.R2BucketEventNotification = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Workers R2 Storage Read`
 * - `Workers R2 Storage Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleR2BucketEventNotification = new cloudflare.R2BucketEventNotification("example_r2_bucket_event_notification", {
 *     accountId: "023e105f4ecef8ad9ca31a8372d0c353",
 *     bucketName: "example-bucket",
 *     queueId: "queue_id",
 *     rules: [{
 *         actions: [
 *             "PutObject",
 *             "CopyObject",
 *         ],
 *         description: "Notifications from source bucket to queue",
 *         prefix: "img/",
 *         suffix: ".jpeg",
 *     }],
 * });
 * ```
 *
 * ## Import
 *
 * > This resource does not currently support `pulumi import`.
 */
class R2BucketEventNotification extends pulumi.CustomResource {
    /**
     * Get an existing R2BucketEventNotification resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new R2BucketEventNotification(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/r2BucketEventNotification:R2BucketEventNotification';
    /**
     * Returns true if the given object is an instance of R2BucketEventNotification.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === R2BucketEventNotification.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["bucketName"] = state?.bucketName;
            resourceInputs["jurisdiction"] = state?.jurisdiction;
            resourceInputs["queueId"] = state?.queueId;
            resourceInputs["queueName"] = state?.queueName;
            resourceInputs["rules"] = state?.rules;
        }
        else {
            const args = argsOrState;
            if (args?.bucketName === undefined && !opts.urn) {
                throw new Error("Missing required property 'bucketName'");
            }
            if (args?.queueId === undefined && !opts.urn) {
                throw new Error("Missing required property 'queueId'");
            }
            if (args?.rules === undefined && !opts.urn) {
                throw new Error("Missing required property 'rules'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["bucketName"] = args?.bucketName;
            resourceInputs["jurisdiction"] = args?.jurisdiction;
            resourceInputs["queueId"] = args?.queueId;
            resourceInputs["rules"] = args?.rules;
            resourceInputs["queueName"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(R2BucketEventNotification.__pulumiType, name, resourceInputs, opts);
    }
}
exports.R2BucketEventNotification = R2BucketEventNotification;
