// @pulumi/cloudflare/spectrumApplication.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpectrumApplication = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Zone Settings Read`
 * - `Zone Settings Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleSpectrumApplication = new cloudflare.SpectrumApplication("example_spectrum_application", {
 *     zoneId: "023e105f4ecef8ad9ca31a8372d0c353",
 *     dns: {
 *         name: "ssh.example.com",
 *         type: "CNAME",
 *     },
 *     protocol: "tcp/22",
 *     trafficType: "direct",
 *     argoSmartRouting: true,
 *     edgeIps: {
 *         connectivity: "all",
 *         type: "dynamic",
 *     },
 *     ipFirewall: false,
 *     originDirects: ["tcp://127.0.0.1:8080"],
 *     originDns: {
 *         name: "origin.example.com",
 *         ttl: 600,
 *         type: "",
 *     },
 *     originPort: 22,
 *     proxyProtocol: "off",
 *     tls: "off",
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/spectrumApplication:SpectrumApplication example '<zone_id>/<app_id>'
 * ```
 */
class SpectrumApplication extends pulumi.CustomResource {
    /**
     * Get an existing SpectrumApplication resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new SpectrumApplication(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/spectrumApplication:SpectrumApplication';
    /**
     * Returns true if the given object is an instance of SpectrumApplication.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === SpectrumApplication.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["argoSmartRouting"] = state?.argoSmartRouting;
            resourceInputs["createdOn"] = state?.createdOn;
            resourceInputs["dns"] = state?.dns;
            resourceInputs["edgeIps"] = state?.edgeIps;
            resourceInputs["ipFirewall"] = state?.ipFirewall;
            resourceInputs["modifiedOn"] = state?.modifiedOn;
            resourceInputs["originDirects"] = state?.originDirects;
            resourceInputs["originDns"] = state?.originDns;
            resourceInputs["originPort"] = state?.originPort;
            resourceInputs["protocol"] = state?.protocol;
            resourceInputs["proxyProtocol"] = state?.proxyProtocol;
            resourceInputs["tls"] = state?.tls;
            resourceInputs["trafficType"] = state?.trafficType;
            resourceInputs["zoneId"] = state?.zoneId;
        }
        else {
            const args = argsOrState;
            if (args?.dns === undefined && !opts.urn) {
                throw new Error("Missing required property 'dns'");
            }
            if (args?.protocol === undefined && !opts.urn) {
                throw new Error("Missing required property 'protocol'");
            }
            resourceInputs["argoSmartRouting"] = args?.argoSmartRouting;
            resourceInputs["dns"] = args?.dns;
            resourceInputs["edgeIps"] = args?.edgeIps;
            resourceInputs["ipFirewall"] = args?.ipFirewall;
            resourceInputs["originDirects"] = args?.originDirects;
            resourceInputs["originDns"] = args?.originDns;
            resourceInputs["originPort"] = args?.originPort;
            resourceInputs["protocol"] = args?.protocol;
            resourceInputs["proxyProtocol"] = args?.proxyProtocol;
            resourceInputs["tls"] = args?.tls;
            resourceInputs["trafficType"] = args?.trafficType;
            resourceInputs["zoneId"] = args?.zoneId;
            resourceInputs["createdOn"] = undefined /*out*/;
            resourceInputs["modifiedOn"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(SpectrumApplication.__pulumiType, name, resourceInputs, opts);
    }
}
exports.SpectrumApplication = SpectrumApplication;
