// @pulumi/cloudflare/teamsAccount.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TeamsAccount = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleZeroTrustGatewaySettings = new cloudflare.ZeroTrustGatewaySettings("example_zero_trust_gateway_settings", {
 *     accountId: "699d98642c564d2e855e9661899b7252",
 *     settings: {
 *         activityLog: {
 *             enabled: true,
 *         },
 *         antivirus: {
 *             enabledDownloadPhase: false,
 *             enabledUploadPhase: false,
 *             failClosed: false,
 *             notificationSettings: {
 *                 enabled: true,
 *                 includeContext: true,
 *                 msg: "msg",
 *                 supportUrl: "support_url",
 *             },
 *         },
 *         blockPage: {
 *             backgroundColor: "background_color",
 *             enabled: true,
 *             footerText: "--footer--",
 *             headerText: "--header--",
 *             includeContext: true,
 *             logoPath: "https://logos.com/a.png",
 *             mailtoAddress: "admin@example.com",
 *             mailtoSubject: "Blocked User Inquiry",
 *             mode: "",
 *             name: "Cloudflare",
 *             suppressFooter: false,
 *             targetUri: "https://example.com",
 *         },
 *         bodyScanning: {
 *             inspectionMode: "deep",
 *         },
 *         browserIsolation: {
 *             nonIdentityEnabled: true,
 *             urlBrowserIsolationEnabled: true,
 *         },
 *         certificate: {
 *             id: "d1b364c5-1311-466e-a194-f0e943e0799f",
 *         },
 *         customCertificate: {
 *             enabled: true,
 *             id: "d1b364c5-1311-466e-a194-f0e943e0799f",
 *         },
 *         extendedEmailMatching: {
 *             enabled: true,
 *         },
 *         fips: {
 *             tls: true,
 *         },
 *         hostSelector: {
 *             enabled: false,
 *         },
 *         inspection: {
 *             mode: "static",
 *         },
 *         protocolDetection: {
 *             enabled: true,
 *         },
 *         sandbox: {
 *             enabled: true,
 *             fallbackAction: "allow",
 *         },
 *         tlsDecrypt: {
 *             enabled: true,
 *         },
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/teamsAccount:TeamsAccount example '<account_id>'
 * ```
 *
 * @deprecated cloudflare.index/teamsaccount.TeamsAccount has been deprecated in favor of cloudflare.index/zerotrustgatewaysettings.ZeroTrustGatewaySettings
 */
class TeamsAccount extends pulumi.CustomResource {
    /**
     * Get an existing TeamsAccount resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        pulumi.log.warn("TeamsAccount is deprecated: cloudflare.index/teamsaccount.TeamsAccount has been deprecated in favor of cloudflare.index/zerotrustgatewaysettings.ZeroTrustGatewaySettings");
        return new TeamsAccount(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/teamsAccount:TeamsAccount';
    /**
     * Returns true if the given object is an instance of TeamsAccount.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === TeamsAccount.__pulumiType;
    }
    /** @deprecated cloudflare.index/teamsaccount.TeamsAccount has been deprecated in favor of cloudflare.index/zerotrustgatewaysettings.ZeroTrustGatewaySettings */
    constructor(name, argsOrState, opts) {
        pulumi.log.warn("TeamsAccount is deprecated: cloudflare.index/teamsaccount.TeamsAccount has been deprecated in favor of cloudflare.index/zerotrustgatewaysettings.ZeroTrustGatewaySettings");
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["createdAt"] = state?.createdAt;
            resourceInputs["settings"] = state?.settings;
            resourceInputs["updatedAt"] = state?.updatedAt;
        }
        else {
            const args = argsOrState;
            if (args?.accountId === undefined && !opts.urn) {
                throw new Error("Missing required property 'accountId'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["settings"] = args?.settings;
            resourceInputs["createdAt"] = undefined /*out*/;
            resourceInputs["updatedAt"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const aliasOpts = { aliases: [{ type: "cloudflare:index/teamsAccount:TeamsAccount" }] };
        opts = pulumi.mergeOptions(opts, aliasOpts);
        super(TeamsAccount.__pulumiType, name, resourceInputs, opts);
    }
}
exports.TeamsAccount = TeamsAccount;
