// @pulumi/cloudflare/turnstileWidget.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TurnstileWidget = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Account Settings Read`
 * - `Account Settings Write`
 * - `Turnstile Sites Read`
 * - `Turnstile Sites Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleTurnstileWidget = new cloudflare.TurnstileWidget("example_turnstile_widget", {
 *     accountId: "023e105f4ecef8ad9ca31a8372d0c353",
 *     domains: [
 *         "203.0.113.1",
 *         "cloudflare.com",
 *         "blog.example.com",
 *     ],
 *     mode: "invisible",
 *     name: "blog.cloudflare.com login form",
 *     botFightMode: false,
 *     clearanceLevel: "interactive",
 *     ephemeralId: false,
 *     offlabel: false,
 *     region: "world",
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/turnstileWidget:TurnstileWidget example '<account_id>/<sitekey>'
 * ```
 */
class TurnstileWidget extends pulumi.CustomResource {
    /**
     * Get an existing TurnstileWidget resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new TurnstileWidget(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/turnstileWidget:TurnstileWidget';
    /**
     * Returns true if the given object is an instance of TurnstileWidget.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === TurnstileWidget.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["botFightMode"] = state?.botFightMode;
            resourceInputs["clearanceLevel"] = state?.clearanceLevel;
            resourceInputs["createdOn"] = state?.createdOn;
            resourceInputs["domains"] = state?.domains;
            resourceInputs["ephemeralId"] = state?.ephemeralId;
            resourceInputs["mode"] = state?.mode;
            resourceInputs["modifiedOn"] = state?.modifiedOn;
            resourceInputs["name"] = state?.name;
            resourceInputs["offlabel"] = state?.offlabel;
            resourceInputs["region"] = state?.region;
            resourceInputs["secret"] = state?.secret;
            resourceInputs["sitekey"] = state?.sitekey;
        }
        else {
            const args = argsOrState;
            if (args?.domains === undefined && !opts.urn) {
                throw new Error("Missing required property 'domains'");
            }
            if (args?.mode === undefined && !opts.urn) {
                throw new Error("Missing required property 'mode'");
            }
            if (args?.name === undefined && !opts.urn) {
                throw new Error("Missing required property 'name'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["botFightMode"] = args?.botFightMode;
            resourceInputs["clearanceLevel"] = args?.clearanceLevel;
            resourceInputs["domains"] = args?.domains;
            resourceInputs["ephemeralId"] = args?.ephemeralId;
            resourceInputs["mode"] = args?.mode;
            resourceInputs["name"] = args?.name;
            resourceInputs["offlabel"] = args?.offlabel;
            resourceInputs["region"] = args?.region;
            resourceInputs["createdOn"] = undefined /*out*/;
            resourceInputs["modifiedOn"] = undefined /*out*/;
            resourceInputs["secret"] = undefined /*out*/;
            resourceInputs["sitekey"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const secretOpts = { additionalSecretOutputs: ["secret"] };
        opts = pulumi.mergeOptions(opts, secretOpts);
        super(TurnstileWidget.__pulumiType, name, resourceInputs, opts);
    }
}
exports.TurnstileWidget = TurnstileWidget;
