// @pulumi/cloudflare/waitingRoom.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WaitingRoom = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Waiting Rooms Read`
 * - `Waiting Rooms Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleWaitingRoom = new cloudflare.WaitingRoom("example_waiting_room", {
 *     zoneId: "023e105f4ecef8ad9ca31a8372d0c353",
 *     host: "shop.example.com",
 *     name: "production_webinar",
 *     newUsersPerMinute: 200,
 *     totalActiveUsers: 200,
 *     additionalRoutes: [{
 *         host: "shop2.example.com",
 *         path: "/shop2/checkout",
 *     }],
 *     cookieAttributes: {
 *         samesite: "auto",
 *         secure: "auto",
 *     },
 *     cookieSuffix: "abcd",
 *     customPageHtml: "{{#waitTimeKnown}} {{waitTime}} mins {{/waitTimeKnown}} {{^waitTimeKnown}} Queue all enabled {{/waitTimeKnown}}",
 *     defaultTemplateLanguage: "es-ES",
 *     description: "Production - DO NOT MODIFY",
 *     disableSessionRenewal: false,
 *     enabledOriginCommands: ["revoke"],
 *     jsonResponseEnabled: false,
 *     path: "/shop/checkout",
 *     queueAll: true,
 *     queueingMethod: "fifo",
 *     queueingStatusCode: 202,
 *     sessionDuration: 1,
 *     suspended: true,
 *     turnstileAction: "log",
 *     turnstileMode: "off",
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/waitingRoom:WaitingRoom example '<zone_id>/<waiting_room_id>'
 * ```
 */
class WaitingRoom extends pulumi.CustomResource {
    /**
     * Get an existing WaitingRoom resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new WaitingRoom(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/waitingRoom:WaitingRoom';
    /**
     * Returns true if the given object is an instance of WaitingRoom.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === WaitingRoom.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["additionalRoutes"] = state?.additionalRoutes;
            resourceInputs["cookieAttributes"] = state?.cookieAttributes;
            resourceInputs["cookieSuffix"] = state?.cookieSuffix;
            resourceInputs["createdOn"] = state?.createdOn;
            resourceInputs["customPageHtml"] = state?.customPageHtml;
            resourceInputs["defaultTemplateLanguage"] = state?.defaultTemplateLanguage;
            resourceInputs["description"] = state?.description;
            resourceInputs["disableSessionRenewal"] = state?.disableSessionRenewal;
            resourceInputs["enabledOriginCommands"] = state?.enabledOriginCommands;
            resourceInputs["host"] = state?.host;
            resourceInputs["jsonResponseEnabled"] = state?.jsonResponseEnabled;
            resourceInputs["modifiedOn"] = state?.modifiedOn;
            resourceInputs["name"] = state?.name;
            resourceInputs["newUsersPerMinute"] = state?.newUsersPerMinute;
            resourceInputs["nextEventPrequeueStartTime"] = state?.nextEventPrequeueStartTime;
            resourceInputs["nextEventStartTime"] = state?.nextEventStartTime;
            resourceInputs["path"] = state?.path;
            resourceInputs["queueAll"] = state?.queueAll;
            resourceInputs["queueingMethod"] = state?.queueingMethod;
            resourceInputs["queueingStatusCode"] = state?.queueingStatusCode;
            resourceInputs["sessionDuration"] = state?.sessionDuration;
            resourceInputs["suspended"] = state?.suspended;
            resourceInputs["totalActiveUsers"] = state?.totalActiveUsers;
            resourceInputs["turnstileAction"] = state?.turnstileAction;
            resourceInputs["turnstileMode"] = state?.turnstileMode;
            resourceInputs["zoneId"] = state?.zoneId;
        }
        else {
            const args = argsOrState;
            if (args?.host === undefined && !opts.urn) {
                throw new Error("Missing required property 'host'");
            }
            if (args?.name === undefined && !opts.urn) {
                throw new Error("Missing required property 'name'");
            }
            if (args?.newUsersPerMinute === undefined && !opts.urn) {
                throw new Error("Missing required property 'newUsersPerMinute'");
            }
            if (args?.totalActiveUsers === undefined && !opts.urn) {
                throw new Error("Missing required property 'totalActiveUsers'");
            }
            resourceInputs["additionalRoutes"] = args?.additionalRoutes;
            resourceInputs["cookieAttributes"] = args?.cookieAttributes;
            resourceInputs["cookieSuffix"] = args?.cookieSuffix;
            resourceInputs["customPageHtml"] = args?.customPageHtml;
            resourceInputs["defaultTemplateLanguage"] = args?.defaultTemplateLanguage;
            resourceInputs["description"] = args?.description;
            resourceInputs["disableSessionRenewal"] = args?.disableSessionRenewal;
            resourceInputs["enabledOriginCommands"] = args?.enabledOriginCommands;
            resourceInputs["host"] = args?.host;
            resourceInputs["jsonResponseEnabled"] = args?.jsonResponseEnabled;
            resourceInputs["name"] = args?.name;
            resourceInputs["newUsersPerMinute"] = args?.newUsersPerMinute;
            resourceInputs["path"] = args?.path;
            resourceInputs["queueAll"] = args?.queueAll;
            resourceInputs["queueingMethod"] = args?.queueingMethod;
            resourceInputs["queueingStatusCode"] = args?.queueingStatusCode;
            resourceInputs["sessionDuration"] = args?.sessionDuration;
            resourceInputs["suspended"] = args?.suspended;
            resourceInputs["totalActiveUsers"] = args?.totalActiveUsers;
            resourceInputs["turnstileAction"] = args?.turnstileAction;
            resourceInputs["turnstileMode"] = args?.turnstileMode;
            resourceInputs["zoneId"] = args?.zoneId;
            resourceInputs["createdOn"] = undefined /*out*/;
            resourceInputs["modifiedOn"] = undefined /*out*/;
            resourceInputs["nextEventPrequeueStartTime"] = undefined /*out*/;
            resourceInputs["nextEventStartTime"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(WaitingRoom.__pulumiType, name, resourceInputs, opts);
    }
}
exports.WaitingRoom = WaitingRoom;
