// @pulumi/cloudflare/workflow.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Workflow = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Workers Scripts Read`
 * - `Workers Scripts Write`
 * - `Workers Tail Read`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleWorkflow = new cloudflare.Workflow("example_workflow", {
 *     accountId: "account_id",
 *     workflowName: "x",
 *     className: "x",
 *     scriptName: "x",
 *     limits: {
 *         steps: 1,
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/workflow:Workflow example '<account_id>/<workflow_name>'
 * ```
 */
class Workflow extends pulumi.CustomResource {
    /**
     * Get an existing Workflow resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Workflow(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/workflow:Workflow';
    /**
     * Returns true if the given object is an instance of Workflow.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Workflow.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["className"] = state?.className;
            resourceInputs["createdOn"] = state?.createdOn;
            resourceInputs["instances"] = state?.instances;
            resourceInputs["isDeleted"] = state?.isDeleted;
            resourceInputs["limits"] = state?.limits;
            resourceInputs["modifiedOn"] = state?.modifiedOn;
            resourceInputs["name"] = state?.name;
            resourceInputs["scriptName"] = state?.scriptName;
            resourceInputs["terminatorRunning"] = state?.terminatorRunning;
            resourceInputs["triggeredOn"] = state?.triggeredOn;
            resourceInputs["versionId"] = state?.versionId;
            resourceInputs["workflowName"] = state?.workflowName;
        }
        else {
            const args = argsOrState;
            if (args?.className === undefined && !opts.urn) {
                throw new Error("Missing required property 'className'");
            }
            if (args?.scriptName === undefined && !opts.urn) {
                throw new Error("Missing required property 'scriptName'");
            }
            if (args?.workflowName === undefined && !opts.urn) {
                throw new Error("Missing required property 'workflowName'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["className"] = args?.className;
            resourceInputs["limits"] = args?.limits;
            resourceInputs["scriptName"] = args?.scriptName;
            resourceInputs["workflowName"] = args?.workflowName;
            resourceInputs["createdOn"] = undefined /*out*/;
            resourceInputs["instances"] = undefined /*out*/;
            resourceInputs["isDeleted"] = undefined /*out*/;
            resourceInputs["modifiedOn"] = undefined /*out*/;
            resourceInputs["name"] = undefined /*out*/;
            resourceInputs["terminatorRunning"] = undefined /*out*/;
            resourceInputs["triggeredOn"] = undefined /*out*/;
            resourceInputs["versionId"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(Workflow.__pulumiType, name, resourceInputs, opts);
    }
}
exports.Workflow = Workflow;
