// @pulumi/cloudflare/workerVersion.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkerVersion = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Workers Scripts Read`
 * - `Workers Scripts Write`
 * - `Workers Tail Read`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleWorkerVersion = new cloudflare.WorkerVersion("example_worker_version", {
 *     accountId: "023e105f4ecef8ad9ca31a8372d0c353",
 *     workerId: "worker_id",
 *     annotations: {
 *         workersMessage: "Fixed bug.",
 *         workersTag: "v1.0.1",
 *     },
 *     assets: {
 *         config: {
 *             htmlHandling: "auto-trailing-slash",
 *             notFoundHandling: "404-page",
 *             runWorkerFirst: [],
 *         },
 *         jwt: "jwt",
 *     },
 *     bindings: [{
 *         name: "MY_ENV_VAR",
 *         text: "my_data",
 *         type: "plain_text",
 *     }],
 *     compatibilityDate: "2021-01-01",
 *     compatibilityFlags: ["nodejs_compat"],
 *     containers: [{
 *         className: "MyDurableObject",
 *     }],
 *     limits: {
 *         cpuMs: 50,
 *         subrequests: 1000,
 *     },
 *     mainModule: "index.js",
 *     migrations: {
 *         deletedClasses: ["string"],
 *         newClasses: ["string"],
 *         newSqliteClasses: ["string"],
 *         newTag: "v2",
 *         oldTag: "v1",
 *         renamedClasses: [{
 *             from: "from",
 *             to: "to",
 *         }],
 *         transferredClasses: [{
 *             from: "from",
 *             fromScript: "from_script",
 *             to: "to",
 *         }],
 *     },
 *     modules: [{
 *         contentFile: "dist/index.js",
 *         contentType: "application/javascript+module",
 *         name: "index.js",
 *     }],
 *     placement: {
 *         mode: "smart",
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/workerVersion:WorkerVersion example '<account_id>/<worker_id>/<version_id>'
 * ```
 */
class WorkerVersion extends pulumi.CustomResource {
    /**
     * Get an existing WorkerVersion resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new WorkerVersion(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/workerVersion:WorkerVersion';
    /**
     * Returns true if the given object is an instance of WorkerVersion.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === WorkerVersion.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["annotations"] = state?.annotations;
            resourceInputs["assets"] = state?.assets;
            resourceInputs["bindings"] = state?.bindings;
            resourceInputs["compatibilityDate"] = state?.compatibilityDate;
            resourceInputs["compatibilityFlags"] = state?.compatibilityFlags;
            resourceInputs["containers"] = state?.containers;
            resourceInputs["createdOn"] = state?.createdOn;
            resourceInputs["limits"] = state?.limits;
            resourceInputs["mainModule"] = state?.mainModule;
            resourceInputs["mainScriptBase64"] = state?.mainScriptBase64;
            resourceInputs["migrationTag"] = state?.migrationTag;
            resourceInputs["migrations"] = state?.migrations;
            resourceInputs["modules"] = state?.modules;
            resourceInputs["number"] = state?.number;
            resourceInputs["placement"] = state?.placement;
            resourceInputs["source"] = state?.source;
            resourceInputs["startupTimeMs"] = state?.startupTimeMs;
            resourceInputs["urls"] = state?.urls;
            resourceInputs["usageModel"] = state?.usageModel;
            resourceInputs["workerId"] = state?.workerId;
        }
        else {
            const args = argsOrState;
            if (args?.workerId === undefined && !opts.urn) {
                throw new Error("Missing required property 'workerId'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["annotations"] = args?.annotations;
            resourceInputs["assets"] = args?.assets;
            resourceInputs["bindings"] = args?.bindings;
            resourceInputs["compatibilityDate"] = args?.compatibilityDate;
            resourceInputs["compatibilityFlags"] = args?.compatibilityFlags;
            resourceInputs["containers"] = args?.containers;
            resourceInputs["limits"] = args?.limits;
            resourceInputs["mainModule"] = args?.mainModule;
            resourceInputs["migrations"] = args?.migrations;
            resourceInputs["modules"] = args?.modules;
            resourceInputs["placement"] = args?.placement;
            resourceInputs["usageModel"] = args?.usageModel;
            resourceInputs["workerId"] = args?.workerId;
            resourceInputs["createdOn"] = undefined /*out*/;
            resourceInputs["mainScriptBase64"] = undefined /*out*/;
            resourceInputs["migrationTag"] = undefined /*out*/;
            resourceInputs["number"] = undefined /*out*/;
            resourceInputs["source"] = undefined /*out*/;
            resourceInputs["startupTimeMs"] = undefined /*out*/;
            resourceInputs["urls"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(WorkerVersion.__pulumiType, name, resourceInputs, opts);
    }
}
exports.WorkerVersion = WorkerVersion;
