// @pulumi/cloudflare/worker.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Worker = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Workers Scripts Read`
 * - `Workers Scripts Write`
 * - `Workers Tail Read`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleWorker = new cloudflare.Worker("example_worker", {
 *     accountId: "023e105f4ecef8ad9ca31a8372d0c353",
 *     name: "my-worker",
 *     logpush: true,
 *     observability: {
 *         enabled: true,
 *         headSamplingRate: 1,
 *         logs: {
 *             destinations: ["string"],
 *             enabled: true,
 *             headSamplingRate: 1,
 *             invocationLogs: true,
 *             persist: true,
 *         },
 *         traces: {
 *             destinations: ["string"],
 *             enabled: true,
 *             headSamplingRate: 1,
 *             persist: true,
 *         },
 *     },
 *     subdomain: {
 *         enabled: true,
 *         previewsEnabled: true,
 *     },
 *     tags: [
 *         "my-team",
 *         "my-public-api",
 *     ],
 *     tailConsumers: [{
 *         name: "my-tail-consumer",
 *     }],
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/worker:Worker example '<account_id>/<worker_id>'
 * ```
 */
class Worker extends pulumi.CustomResource {
    /**
     * Get an existing Worker resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Worker(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/worker:Worker';
    /**
     * Returns true if the given object is an instance of Worker.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Worker.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["createdOn"] = state?.createdOn;
            resourceInputs["deployedOn"] = state?.deployedOn;
            resourceInputs["logpush"] = state?.logpush;
            resourceInputs["name"] = state?.name;
            resourceInputs["observability"] = state?.observability;
            resourceInputs["references"] = state?.references;
            resourceInputs["subdomain"] = state?.subdomain;
            resourceInputs["tags"] = state?.tags;
            resourceInputs["tailConsumers"] = state?.tailConsumers;
            resourceInputs["updatedOn"] = state?.updatedOn;
        }
        else {
            const args = argsOrState;
            if (args?.name === undefined && !opts.urn) {
                throw new Error("Missing required property 'name'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["logpush"] = args?.logpush;
            resourceInputs["name"] = args?.name;
            resourceInputs["observability"] = args?.observability;
            resourceInputs["subdomain"] = args?.subdomain;
            resourceInputs["tags"] = args?.tags;
            resourceInputs["tailConsumers"] = args?.tailConsumers;
            resourceInputs["createdOn"] = undefined /*out*/;
            resourceInputs["deployedOn"] = undefined /*out*/;
            resourceInputs["references"] = undefined /*out*/;
            resourceInputs["updatedOn"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(Worker.__pulumiType, name, resourceInputs, opts);
    }
}
exports.Worker = Worker;
