// @pulumi/cloudflare/workersScript.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkersScript = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Workers Scripts Read`
 * - `Workers Scripts Write`
 * - `Workers Tail Read`
 *
 * > For more direct control over Workers resources, we recommend the beta `cloudflare.Worker`, `cloudflare.WorkerVersion`, and `cloudflare.WorkersDeployment` resources. See how to use them in the [developer documentation](https://developers.cloudflare.com/workers/platform/infrastructure-as-code/).
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/workersScript:WorkersScript example '<account_id>/<script_name>'
 * ```
 */
class WorkersScript extends pulumi.CustomResource {
    /**
     * Get an existing WorkersScript resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new WorkersScript(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/workersScript:WorkersScript';
    /**
     * Returns true if the given object is an instance of WorkersScript.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === WorkersScript.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["annotations"] = state?.annotations;
            resourceInputs["assets"] = state?.assets;
            resourceInputs["bindings"] = state?.bindings;
            resourceInputs["bodyPart"] = state?.bodyPart;
            resourceInputs["compatibilityDate"] = state?.compatibilityDate;
            resourceInputs["compatibilityFlags"] = state?.compatibilityFlags;
            resourceInputs["content"] = state?.content;
            resourceInputs["contentFile"] = state?.contentFile;
            resourceInputs["contentSha256"] = state?.contentSha256;
            resourceInputs["contentType"] = state?.contentType;
            resourceInputs["createdOn"] = state?.createdOn;
            resourceInputs["etag"] = state?.etag;
            resourceInputs["handlers"] = state?.handlers;
            resourceInputs["hasAssets"] = state?.hasAssets;
            resourceInputs["hasModules"] = state?.hasModules;
            resourceInputs["keepAssets"] = state?.keepAssets;
            resourceInputs["keepBindings"] = state?.keepBindings;
            resourceInputs["lastDeployedFrom"] = state?.lastDeployedFrom;
            resourceInputs["limits"] = state?.limits;
            resourceInputs["logpush"] = state?.logpush;
            resourceInputs["mainModule"] = state?.mainModule;
            resourceInputs["migrationTag"] = state?.migrationTag;
            resourceInputs["migrations"] = state?.migrations;
            resourceInputs["modifiedOn"] = state?.modifiedOn;
            resourceInputs["namedHandlers"] = state?.namedHandlers;
            resourceInputs["observability"] = state?.observability;
            resourceInputs["placement"] = state?.placement;
            resourceInputs["placementMode"] = state?.placementMode;
            resourceInputs["placementStatus"] = state?.placementStatus;
            resourceInputs["scriptName"] = state?.scriptName;
            resourceInputs["startupTimeMs"] = state?.startupTimeMs;
            resourceInputs["tailConsumers"] = state?.tailConsumers;
            resourceInputs["usageModel"] = state?.usageModel;
        }
        else {
            const args = argsOrState;
            if (args?.scriptName === undefined && !opts.urn) {
                throw new Error("Missing required property 'scriptName'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["annotations"] = args?.annotations;
            resourceInputs["assets"] = args?.assets;
            resourceInputs["bindings"] = args?.bindings;
            resourceInputs["bodyPart"] = args?.bodyPart;
            resourceInputs["compatibilityDate"] = args?.compatibilityDate;
            resourceInputs["compatibilityFlags"] = args?.compatibilityFlags;
            resourceInputs["content"] = args?.content;
            resourceInputs["contentFile"] = args?.contentFile;
            resourceInputs["contentSha256"] = args?.contentSha256;
            resourceInputs["contentType"] = args?.contentType;
            resourceInputs["keepAssets"] = args?.keepAssets;
            resourceInputs["keepBindings"] = args?.keepBindings;
            resourceInputs["limits"] = args?.limits;
            resourceInputs["logpush"] = args?.logpush;
            resourceInputs["mainModule"] = args?.mainModule;
            resourceInputs["migrations"] = args?.migrations;
            resourceInputs["observability"] = args?.observability;
            resourceInputs["placement"] = args?.placement;
            resourceInputs["scriptName"] = args?.scriptName;
            resourceInputs["tailConsumers"] = args?.tailConsumers;
            resourceInputs["usageModel"] = args?.usageModel;
            resourceInputs["createdOn"] = undefined /*out*/;
            resourceInputs["etag"] = undefined /*out*/;
            resourceInputs["handlers"] = undefined /*out*/;
            resourceInputs["hasAssets"] = undefined /*out*/;
            resourceInputs["hasModules"] = undefined /*out*/;
            resourceInputs["lastDeployedFrom"] = undefined /*out*/;
            resourceInputs["migrationTag"] = undefined /*out*/;
            resourceInputs["modifiedOn"] = undefined /*out*/;
            resourceInputs["namedHandlers"] = undefined /*out*/;
            resourceInputs["placementMode"] = undefined /*out*/;
            resourceInputs["placementStatus"] = undefined /*out*/;
            resourceInputs["startupTimeMs"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const aliasOpts = { aliases: [{ type: "cloudflare:index/workerScript:WorkerScript" }] };
        opts = pulumi.mergeOptions(opts, aliasOpts);
        super(WorkersScript.__pulumiType, name, resourceInputs, opts);
    }
}
exports.WorkersScript = WorkersScript;
