// @pulumi/cloudflare/zeroTrustDnsLocation.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZeroTrustDnsLocation = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Cloudflare Zero Trust Secure DNS Locations Write`
 * - `Zero Trust Read`
 * - `Zero Trust Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleZeroTrustDnsLocation = new cloudflare.ZeroTrustDnsLocation("example_zero_trust_dns_location", {
 *     accountId: "699d98642c564d2e855e9661899b7252",
 *     name: "Austin Office Location",
 *     clientDefault: false,
 *     dnsDestinationIpsId: "0e4a32c6-6fb8-4858-9296-98f51631e8e6",
 *     ecsSupport: false,
 *     endpoints: {
 *         doh: {
 *             enabled: true,
 *             networks: [{
 *                 network: "2001:85a3::/64",
 *             }],
 *             requireToken: true,
 *         },
 *         dot: {
 *             enabled: true,
 *             networks: [{
 *                 network: "2001:85a3::/64",
 *             }],
 *         },
 *         ipv4: {
 *             enabled: true,
 *         },
 *         ipv6: {
 *             enabled: true,
 *             networks: [{
 *                 network: "2001:85a3::/64",
 *             }],
 *         },
 *     },
 *     networks: [{
 *         network: "192.0.2.1/32",
 *     }],
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/zeroTrustDnsLocation:ZeroTrustDnsLocation example '<account_id>/<location_id>'
 * ```
 */
class ZeroTrustDnsLocation extends pulumi.CustomResource {
    /**
     * Get an existing ZeroTrustDnsLocation resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ZeroTrustDnsLocation(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/zeroTrustDnsLocation:ZeroTrustDnsLocation';
    /**
     * Returns true if the given object is an instance of ZeroTrustDnsLocation.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ZeroTrustDnsLocation.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["clientDefault"] = state?.clientDefault;
            resourceInputs["createdAt"] = state?.createdAt;
            resourceInputs["dnsDestinationIpsId"] = state?.dnsDestinationIpsId;
            resourceInputs["dnsDestinationIpv6BlockId"] = state?.dnsDestinationIpv6BlockId;
            resourceInputs["dohSubdomain"] = state?.dohSubdomain;
            resourceInputs["ecsSupport"] = state?.ecsSupport;
            resourceInputs["endpoints"] = state?.endpoints;
            resourceInputs["ip"] = state?.ip;
            resourceInputs["ipv4Destination"] = state?.ipv4Destination;
            resourceInputs["ipv4DestinationBackup"] = state?.ipv4DestinationBackup;
            resourceInputs["name"] = state?.name;
            resourceInputs["networks"] = state?.networks;
            resourceInputs["updatedAt"] = state?.updatedAt;
        }
        else {
            const args = argsOrState;
            if (args?.name === undefined && !opts.urn) {
                throw new Error("Missing required property 'name'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["clientDefault"] = args?.clientDefault;
            resourceInputs["dnsDestinationIpsId"] = args?.dnsDestinationIpsId;
            resourceInputs["ecsSupport"] = args?.ecsSupport;
            resourceInputs["endpoints"] = args?.endpoints;
            resourceInputs["name"] = args?.name;
            resourceInputs["networks"] = args?.networks;
            resourceInputs["createdAt"] = undefined /*out*/;
            resourceInputs["dnsDestinationIpv6BlockId"] = undefined /*out*/;
            resourceInputs["dohSubdomain"] = undefined /*out*/;
            resourceInputs["ip"] = undefined /*out*/;
            resourceInputs["ipv4Destination"] = undefined /*out*/;
            resourceInputs["ipv4DestinationBackup"] = undefined /*out*/;
            resourceInputs["updatedAt"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const aliasOpts = { aliases: [{ type: "cloudflare:index/teamsLocation:TeamsLocation" }] };
        opts = pulumi.mergeOptions(opts, aliasOpts);
        super(ZeroTrustDnsLocation.__pulumiType, name, resourceInputs, opts);
    }
}
exports.ZeroTrustDnsLocation = ZeroTrustDnsLocation;
