// @pulumi/cloudflare/zeroTrustAccessInfrastructureTarget.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZeroTrustAccessInfrastructureTarget = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleZeroTrustAccessInfrastructureTarget = new cloudflare.ZeroTrustAccessInfrastructureTarget("example_zero_trust_access_infrastructure_target", {
 *     accountId: "023e105f4ecef8ad9ca31a8372d0c353",
 *     hostname: "infra-access-target",
 *     ip: {
 *         ipv4: {
 *             ipAddr: "187.26.29.249",
 *             virtualNetworkId: "c77b744e-acc8-428f-9257-6878c046ed55",
 *         },
 *         ipv6: {
 *             ipAddr: "64c0:64e8:f0b4:8dbf:7104:72b0:ec8f:f5e0",
 *             virtualNetworkId: "c77b744e-acc8-428f-9257-6878c046ed55",
 *         },
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/zeroTrustAccessInfrastructureTarget:ZeroTrustAccessInfrastructureTarget example '<account_id>/<target_id>'
 * ```
 */
class ZeroTrustAccessInfrastructureTarget extends pulumi.CustomResource {
    /**
     * Get an existing ZeroTrustAccessInfrastructureTarget resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ZeroTrustAccessInfrastructureTarget(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/zeroTrustAccessInfrastructureTarget:ZeroTrustAccessInfrastructureTarget';
    /**
     * Returns true if the given object is an instance of ZeroTrustAccessInfrastructureTarget.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ZeroTrustAccessInfrastructureTarget.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["createdAt"] = state?.createdAt;
            resourceInputs["hostname"] = state?.hostname;
            resourceInputs["ip"] = state?.ip;
            resourceInputs["modifiedAt"] = state?.modifiedAt;
        }
        else {
            const args = argsOrState;
            if (args?.hostname === undefined && !opts.urn) {
                throw new Error("Missing required property 'hostname'");
            }
            if (args?.ip === undefined && !opts.urn) {
                throw new Error("Missing required property 'ip'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["hostname"] = args?.hostname;
            resourceInputs["ip"] = args?.ip;
            resourceInputs["createdAt"] = undefined /*out*/;
            resourceInputs["modifiedAt"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(ZeroTrustAccessInfrastructureTarget.__pulumiType, name, resourceInputs, opts);
    }
}
exports.ZeroTrustAccessInfrastructureTarget = ZeroTrustAccessInfrastructureTarget;
