// @pulumi/cloudflare/zeroTrustTunnelCloudflared.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZeroTrustTunnelCloudflared = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Cloudflare One Connector: cloudflared Read`
 * - `Cloudflare One Connector: cloudflared Write`
 * - `Cloudflare One Connectors Read`
 * - `Cloudflare One Connectors Write`
 * - `Cloudflare Tunnel Read`
 * - `Cloudflare Tunnel Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleZeroTrustTunnelCloudflared = new cloudflare.ZeroTrustTunnelCloudflared("example_zero_trust_tunnel_cloudflared", {
 *     accountId: "699d98642c564d2e855e9661899b7252",
 *     name: "blog",
 *     configSrc: "cloudflare",
 *     tunnelSecret: "AQIDBAUGBwgBAgMEBQYHCAECAwQFBgcIAQIDBAUGBwg=",
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/zeroTrustTunnelCloudflared:ZeroTrustTunnelCloudflared example '<account_id>/<tunnel_id>'
 * ```
 */
class ZeroTrustTunnelCloudflared extends pulumi.CustomResource {
    /**
     * Get an existing ZeroTrustTunnelCloudflared resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ZeroTrustTunnelCloudflared(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/zeroTrustTunnelCloudflared:ZeroTrustTunnelCloudflared';
    /**
     * Returns true if the given object is an instance of ZeroTrustTunnelCloudflared.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ZeroTrustTunnelCloudflared.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["accountTag"] = state?.accountTag;
            resourceInputs["configSrc"] = state?.configSrc;
            resourceInputs["connections"] = state?.connections;
            resourceInputs["connsActiveAt"] = state?.connsActiveAt;
            resourceInputs["connsInactiveAt"] = state?.connsInactiveAt;
            resourceInputs["createdAt"] = state?.createdAt;
            resourceInputs["deletedAt"] = state?.deletedAt;
            resourceInputs["metadata"] = state?.metadata;
            resourceInputs["name"] = state?.name;
            resourceInputs["remoteConfig"] = state?.remoteConfig;
            resourceInputs["status"] = state?.status;
            resourceInputs["tunType"] = state?.tunType;
            resourceInputs["tunnelSecret"] = state?.tunnelSecret;
        }
        else {
            const args = argsOrState;
            if (args?.name === undefined && !opts.urn) {
                throw new Error("Missing required property 'name'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["configSrc"] = args?.configSrc;
            resourceInputs["name"] = args?.name;
            resourceInputs["tunnelSecret"] = args?.tunnelSecret ? pulumi.secret(args.tunnelSecret) : undefined;
            resourceInputs["accountTag"] = undefined /*out*/;
            resourceInputs["connections"] = undefined /*out*/;
            resourceInputs["connsActiveAt"] = undefined /*out*/;
            resourceInputs["connsInactiveAt"] = undefined /*out*/;
            resourceInputs["createdAt"] = undefined /*out*/;
            resourceInputs["deletedAt"] = undefined /*out*/;
            resourceInputs["metadata"] = undefined /*out*/;
            resourceInputs["remoteConfig"] = undefined /*out*/;
            resourceInputs["status"] = undefined /*out*/;
            resourceInputs["tunType"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const aliasOpts = { aliases: [{ type: "cloudflare:index/tunnel:Tunnel" }] };
        opts = pulumi.mergeOptions(opts, aliasOpts);
        const secretOpts = { additionalSecretOutputs: ["tunnelSecret"] };
        opts = pulumi.mergeOptions(opts, secretOpts);
        super(ZeroTrustTunnelCloudflared.__pulumiType, name, resourceInputs, opts);
    }
}
exports.ZeroTrustTunnelCloudflared = ZeroTrustTunnelCloudflared;
