// @pulumi/cloudflare/zeroTrustAccessApplication.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZeroTrustAccessApplication = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * ## Example Usage
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/zeroTrustAccessApplication:ZeroTrustAccessApplication example '<{accounts|zones}/{account_id|zone_id}>/<app_id>'
 * ```
 */
class ZeroTrustAccessApplication extends pulumi.CustomResource {
    /**
     * Get an existing ZeroTrustAccessApplication resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ZeroTrustAccessApplication(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/zeroTrustAccessApplication:ZeroTrustAccessApplication';
    /**
     * Returns true if the given object is an instance of ZeroTrustAccessApplication.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ZeroTrustAccessApplication.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["allowAuthenticateViaWarp"] = state?.allowAuthenticateViaWarp;
            resourceInputs["allowIframe"] = state?.allowIframe;
            resourceInputs["allowedIdps"] = state?.allowedIdps;
            resourceInputs["appLauncherLogoUrl"] = state?.appLauncherLogoUrl;
            resourceInputs["appLauncherVisible"] = state?.appLauncherVisible;
            resourceInputs["aud"] = state?.aud;
            resourceInputs["autoRedirectToIdentity"] = state?.autoRedirectToIdentity;
            resourceInputs["bgColor"] = state?.bgColor;
            resourceInputs["corsHeaders"] = state?.corsHeaders;
            resourceInputs["customDenyMessage"] = state?.customDenyMessage;
            resourceInputs["customDenyUrl"] = state?.customDenyUrl;
            resourceInputs["customNonIdentityDenyUrl"] = state?.customNonIdentityDenyUrl;
            resourceInputs["customPages"] = state?.customPages;
            resourceInputs["destinations"] = state?.destinations;
            resourceInputs["domain"] = state?.domain;
            resourceInputs["enableBindingCookie"] = state?.enableBindingCookie;
            resourceInputs["footerLinks"] = state?.footerLinks;
            resourceInputs["headerBgColor"] = state?.headerBgColor;
            resourceInputs["httpOnlyCookieAttribute"] = state?.httpOnlyCookieAttribute;
            resourceInputs["landingPageDesign"] = state?.landingPageDesign;
            resourceInputs["logoUrl"] = state?.logoUrl;
            resourceInputs["mfaConfig"] = state?.mfaConfig;
            resourceInputs["name"] = state?.name;
            resourceInputs["oauthConfiguration"] = state?.oauthConfiguration;
            resourceInputs["optionsPreflightBypass"] = state?.optionsPreflightBypass;
            resourceInputs["pathCookieAttribute"] = state?.pathCookieAttribute;
            resourceInputs["policies"] = state?.policies;
            resourceInputs["readServiceTokensFromHeader"] = state?.readServiceTokensFromHeader;
            resourceInputs["saasApp"] = state?.saasApp;
            resourceInputs["sameSiteCookieAttribute"] = state?.sameSiteCookieAttribute;
            resourceInputs["scimConfig"] = state?.scimConfig;
            resourceInputs["selfHostedDomains"] = state?.selfHostedDomains;
            resourceInputs["serviceAuth401Redirect"] = state?.serviceAuth401Redirect;
            resourceInputs["sessionDuration"] = state?.sessionDuration;
            resourceInputs["skipAppLauncherLoginPage"] = state?.skipAppLauncherLoginPage;
            resourceInputs["skipInterstitial"] = state?.skipInterstitial;
            resourceInputs["tags"] = state?.tags;
            resourceInputs["targetCriterias"] = state?.targetCriterias;
            resourceInputs["type"] = state?.type;
            resourceInputs["zoneId"] = state?.zoneId;
        }
        else {
            const args = argsOrState;
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["allowAuthenticateViaWarp"] = args?.allowAuthenticateViaWarp;
            resourceInputs["allowIframe"] = args?.allowIframe;
            resourceInputs["allowedIdps"] = args?.allowedIdps;
            resourceInputs["appLauncherLogoUrl"] = args?.appLauncherLogoUrl;
            resourceInputs["appLauncherVisible"] = args?.appLauncherVisible;
            resourceInputs["autoRedirectToIdentity"] = args?.autoRedirectToIdentity;
            resourceInputs["bgColor"] = args?.bgColor;
            resourceInputs["corsHeaders"] = args?.corsHeaders;
            resourceInputs["customDenyMessage"] = args?.customDenyMessage;
            resourceInputs["customDenyUrl"] = args?.customDenyUrl;
            resourceInputs["customNonIdentityDenyUrl"] = args?.customNonIdentityDenyUrl;
            resourceInputs["customPages"] = args?.customPages;
            resourceInputs["destinations"] = args?.destinations;
            resourceInputs["domain"] = args?.domain;
            resourceInputs["enableBindingCookie"] = args?.enableBindingCookie;
            resourceInputs["footerLinks"] = args?.footerLinks;
            resourceInputs["headerBgColor"] = args?.headerBgColor;
            resourceInputs["httpOnlyCookieAttribute"] = args?.httpOnlyCookieAttribute;
            resourceInputs["landingPageDesign"] = args?.landingPageDesign;
            resourceInputs["logoUrl"] = args?.logoUrl;
            resourceInputs["mfaConfig"] = args?.mfaConfig;
            resourceInputs["name"] = args?.name;
            resourceInputs["oauthConfiguration"] = args?.oauthConfiguration;
            resourceInputs["optionsPreflightBypass"] = args?.optionsPreflightBypass;
            resourceInputs["pathCookieAttribute"] = args?.pathCookieAttribute;
            resourceInputs["policies"] = args?.policies;
            resourceInputs["readServiceTokensFromHeader"] = args?.readServiceTokensFromHeader;
            resourceInputs["saasApp"] = args?.saasApp;
            resourceInputs["sameSiteCookieAttribute"] = args?.sameSiteCookieAttribute;
            resourceInputs["scimConfig"] = args?.scimConfig;
            resourceInputs["selfHostedDomains"] = args?.selfHostedDomains;
            resourceInputs["serviceAuth401Redirect"] = args?.serviceAuth401Redirect;
            resourceInputs["sessionDuration"] = args?.sessionDuration;
            resourceInputs["skipAppLauncherLoginPage"] = args?.skipAppLauncherLoginPage;
            resourceInputs["skipInterstitial"] = args?.skipInterstitial;
            resourceInputs["tags"] = args?.tags;
            resourceInputs["targetCriterias"] = args?.targetCriterias;
            resourceInputs["type"] = args?.type;
            resourceInputs["zoneId"] = args?.zoneId;
            resourceInputs["aud"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const aliasOpts = { aliases: [{ type: "cloudflare:index/accessApplication:AccessApplication" }] };
        opts = pulumi.mergeOptions(opts, aliasOpts);
        super(ZeroTrustAccessApplication.__pulumiType, name, resourceInputs, opts);
    }
}
exports.ZeroTrustAccessApplication = ZeroTrustAccessApplication;
