// @pulumi/cloudflare/zeroTrustGatewayPolicy.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZeroTrustGatewayPolicy = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleZeroTrustGatewayPolicy = new cloudflare.ZeroTrustGatewayPolicy("example_zero_trust_gateway_policy", {
 *     accountId: "699d98642c564d2e855e9661899b7252",
 *     action: "allow",
 *     name: "block bad websites",
 *     description: "Block bad websites based on their host name.",
 *     devicePosture: "any(device_posture.checks.passed[*] in {\"1308749e-fcfb-4ebc-b051-fe022b632644\"})",
 *     enabled: true,
 *     expiration: {
 *         expiresAt: "2014-01-01T05:20:20Z",
 *         duration: 10,
 *     },
 *     filters: ["http"],
 *     identity: "any(identity.groups.name[*] in {\"finance\"})",
 *     precedence: 0,
 *     ruleSettings: {
 *         addHeaders: {
 *             "My-Next-Header": [
 *                 "foo",
 *                 "bar",
 *             ],
 *             "X-Custom-Header-Name": ["somecustomvalue"],
 *         },
 *         allowChildBypass: false,
 *         auditSsh: {
 *             commandLogging: false,
 *         },
 *         bisoAdminControls: {
 *             copy: "remote_only",
 *             dcp: true,
 *             dd: true,
 *             dk: true,
 *             download: "enabled",
 *             dp: false,
 *             du: true,
 *             keyboard: "enabled",
 *             paste: "enabled",
 *             printing: "enabled",
 *             upload: "enabled",
 *             version: "v1",
 *             wmId: "475345dc-5299-4b6e-8f6a-3d3e4c8e9f1a",
 *         },
 *         blockPage: {
 *             targetUri: "https://example.com",
 *             includeContext: true,
 *         },
 *         blockPageEnabled: true,
 *         blockReason: "This website is a security risk",
 *         bypassParentRule: false,
 *         checkSession: {
 *             duration: "300s",
 *             enforce: true,
 *         },
 *         dnsResolvers: {
 *             ipv4s: [{
 *                 ip: "2.2.2.2",
 *                 port: 5053,
 *                 routeThroughPrivateNetwork: true,
 *                 vnetId: "f174e90a-fafe-4643-bbbc-4a0ed4fc8415",
 *             }],
 *             ipv6s: [{
 *                 ip: "2001:DB8::",
 *                 port: 5053,
 *                 routeThroughPrivateNetwork: true,
 *                 vnetId: "f174e90a-fafe-4643-bbbc-4a0ed4fc8415",
 *             }],
 *         },
 *         egress: {
 *             ipv4: "192.0.2.2",
 *             ipv4Fallback: "192.0.2.3",
 *             ipv6: "2001:DB8::/64",
 *         },
 *         forensicCopy: {
 *             enabled: true,
 *         },
 *         ignoreCnameCategoryMatches: true,
 *         insecureDisableDnssecValidation: false,
 *         ipCategories: true,
 *         ipIndicatorFeeds: true,
 *         l4override: {
 *             ip: "1.1.1.1",
 *             port: 0,
 *         },
 *         notificationSettings: {
 *             enabled: true,
 *             includeContext: true,
 *             msg: "msg",
 *             supportUrl: "support_url",
 *         },
 *         overrideHost: "example.com",
 *         overrideIps: [
 *             "1.1.1.1",
 *             "2.2.2.2",
 *         ],
 *         payloadLog: {
 *             enabled: true,
 *         },
 *         quarantine: {
 *             fileTypes: ["exe"],
 *         },
 *         redirect: {
 *             targetUri: "https://example.com",
 *             includeContext: true,
 *             preservePathAndQuery: true,
 *         },
 *         resolveDnsInternally: {
 *             fallback: "none",
 *             viewId: "view_id",
 *         },
 *         resolveDnsThroughCloudflare: true,
 *         untrustedCert: {
 *             action: "error",
 *         },
 *     },
 *     schedule: {
 *         fri: "08:00-12:30,13:30-17:00",
 *         mon: "08:00-12:30,13:30-17:00",
 *         sat: "08:00-12:30,13:30-17:00",
 *         sun: "08:00-12:30,13:30-17:00",
 *         thu: "08:00-12:30,13:30-17:00",
 *         timeZone: "America/New York",
 *         tue: "08:00-12:30,13:30-17:00",
 *         wed: "08:00-12:30,13:30-17:00",
 *     },
 *     traffic: "http.request.uri matches \".*a/partial/uri.*\" and http.request.host in $01302951-49f9-47c9-a400-0297e60b6a10",
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/zeroTrustGatewayPolicy:ZeroTrustGatewayPolicy example '<account_id>/<rule_id>'
 * ```
 */
class ZeroTrustGatewayPolicy extends pulumi.CustomResource {
    /**
     * Get an existing ZeroTrustGatewayPolicy resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ZeroTrustGatewayPolicy(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/zeroTrustGatewayPolicy:ZeroTrustGatewayPolicy';
    /**
     * Returns true if the given object is an instance of ZeroTrustGatewayPolicy.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ZeroTrustGatewayPolicy.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["action"] = state?.action;
            resourceInputs["createdAt"] = state?.createdAt;
            resourceInputs["deletedAt"] = state?.deletedAt;
            resourceInputs["description"] = state?.description;
            resourceInputs["devicePosture"] = state?.devicePosture;
            resourceInputs["enabled"] = state?.enabled;
            resourceInputs["expiration"] = state?.expiration;
            resourceInputs["filters"] = state?.filters;
            resourceInputs["identity"] = state?.identity;
            resourceInputs["name"] = state?.name;
            resourceInputs["precedence"] = state?.precedence;
            resourceInputs["readOnly"] = state?.readOnly;
            resourceInputs["ruleSettings"] = state?.ruleSettings;
            resourceInputs["schedule"] = state?.schedule;
            resourceInputs["sharable"] = state?.sharable;
            resourceInputs["sourceAccount"] = state?.sourceAccount;
            resourceInputs["traffic"] = state?.traffic;
            resourceInputs["updatedAt"] = state?.updatedAt;
            resourceInputs["version"] = state?.version;
            resourceInputs["warningStatus"] = state?.warningStatus;
        }
        else {
            const args = argsOrState;
            if (args?.action === undefined && !opts.urn) {
                throw new Error("Missing required property 'action'");
            }
            if (args?.name === undefined && !opts.urn) {
                throw new Error("Missing required property 'name'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["action"] = args?.action;
            resourceInputs["description"] = args?.description;
            resourceInputs["devicePosture"] = args?.devicePosture;
            resourceInputs["enabled"] = args?.enabled;
            resourceInputs["expiration"] = args?.expiration;
            resourceInputs["filters"] = args?.filters;
            resourceInputs["identity"] = args?.identity;
            resourceInputs["name"] = args?.name;
            resourceInputs["precedence"] = args?.precedence;
            resourceInputs["ruleSettings"] = args?.ruleSettings;
            resourceInputs["schedule"] = args?.schedule;
            resourceInputs["traffic"] = args?.traffic;
            resourceInputs["createdAt"] = undefined /*out*/;
            resourceInputs["deletedAt"] = undefined /*out*/;
            resourceInputs["readOnly"] = undefined /*out*/;
            resourceInputs["sharable"] = undefined /*out*/;
            resourceInputs["sourceAccount"] = undefined /*out*/;
            resourceInputs["updatedAt"] = undefined /*out*/;
            resourceInputs["version"] = undefined /*out*/;
            resourceInputs["warningStatus"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const aliasOpts = { aliases: [{ type: "cloudflare:index/teamsRule:TeamsRule" }] };
        opts = pulumi.mergeOptions(opts, aliasOpts);
        super(ZeroTrustGatewayPolicy.__pulumiType, name, resourceInputs, opts);
    }
}
exports.ZeroTrustGatewayPolicy = ZeroTrustGatewayPolicy;
