// @pulumi/cloudflare/zeroTrustDlpPredefinedProfile.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZeroTrustDlpPredefinedProfile = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Zero Trust Read`
 * - `Zero Trust Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleZeroTrustDlpPredefinedProfile = new cloudflare.ZeroTrustDlpPredefinedProfile("example_zero_trust_dlp_predefined_profile", {
 *     profileId: "e91a2360-da51-4fdf-9711-bcdecd462614",
 *     accountId: "account_id",
 *     ocrEnabled: true,
 *     enabledEntries: [
 *         "56a8c060-01bb-4f89-ba1e-3ad42770a342",
 *         "7f575e6d-039a-465e-85cf-175bda88d4f2",
 *         "03ebabfd-ce7e-45ed-8061-65e28f0a6e53",
 *         "2d9c356d-b5a3-482a-b01e-0363e0de7458",
 *         "2f3657af-c39b-4899-9a98-22f7d187dd28",
 *         "753a16f9-f533-4208-a5b8-6319b201e9fb",
 *         "ebcea2c4-335a-457c-853b-f7ae7cc74e07",
 *         "3f5c4c83-f34c-4d17-81c7-3028385737b3",
 *         "d1a84fde-c375-4d3c-8a27-8c4eaa33cf60",
 *         "6dbe5604-d3a3-4c3e-905c-57985704bea7",
 *         "55ba2c6c-8ef4-4b2e-9148-e75e8b6ccac1",
 *         "5b1d5035-8c53-4bc9-a151-404eb32b34b4",
 *         "acf28d88-2daf-4bc4-aa36-5ac1fac0540a",
 *     ],
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/zeroTrustDlpPredefinedProfile:ZeroTrustDlpPredefinedProfile example '<account_id>/<profile_id>'
 * ```
 */
class ZeroTrustDlpPredefinedProfile extends pulumi.CustomResource {
    /**
     * Get an existing ZeroTrustDlpPredefinedProfile resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ZeroTrustDlpPredefinedProfile(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/zeroTrustDlpPredefinedProfile:ZeroTrustDlpPredefinedProfile';
    /**
     * Returns true if the given object is an instance of ZeroTrustDlpPredefinedProfile.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ZeroTrustDlpPredefinedProfile.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["aiContextEnabled"] = state?.aiContextEnabled;
            resourceInputs["allowedMatchCount"] = state?.allowedMatchCount;
            resourceInputs["confidenceThreshold"] = state?.confidenceThreshold;
            resourceInputs["enabledEntries"] = state?.enabledEntries;
            resourceInputs["entries"] = state?.entries;
            resourceInputs["name"] = state?.name;
            resourceInputs["ocrEnabled"] = state?.ocrEnabled;
            resourceInputs["openAccess"] = state?.openAccess;
            resourceInputs["profileId"] = state?.profileId;
        }
        else {
            const args = argsOrState;
            if (args?.profileId === undefined && !opts.urn) {
                throw new Error("Missing required property 'profileId'");
            }
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["aiContextEnabled"] = args?.aiContextEnabled;
            resourceInputs["allowedMatchCount"] = args?.allowedMatchCount;
            resourceInputs["confidenceThreshold"] = args?.confidenceThreshold;
            resourceInputs["enabledEntries"] = args?.enabledEntries;
            resourceInputs["entries"] = args?.entries;
            resourceInputs["ocrEnabled"] = args?.ocrEnabled;
            resourceInputs["profileId"] = args?.profileId;
            resourceInputs["name"] = undefined /*out*/;
            resourceInputs["openAccess"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const aliasOpts = { aliases: [{ type: "cloudflare:index/dlpPredefinedProfile:DlpPredefinedProfile" }] };
        opts = pulumi.mergeOptions(opts, aliasOpts);
        super(ZeroTrustDlpPredefinedProfile.__pulumiType, name, resourceInputs, opts);
    }
}
exports.ZeroTrustDlpPredefinedProfile = ZeroTrustDlpPredefinedProfile;
