// @pulumi/cloudflare/zeroTrustDeviceSettings.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZeroTrustDeviceSettings = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Zero Trust Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleZeroTrustDeviceSettings = new cloudflare.ZeroTrustDeviceSettings("example_zero_trust_device_settings", {
 *     accountId: "699d98642c564d2e855e9661899b7252",
 *     disableForTime: 0,
 *     externalEmergencySignalEnabled: true,
 *     externalEmergencySignalFingerprint: "abcd1234567890abcd1234567890abcd1234567890abcd1234567890abcd1234",
 *     externalEmergencySignalInterval: "5m",
 *     externalEmergencySignalUrl: "https://192.0.2.1/signal",
 *     gatewayProxyEnabled: true,
 *     gatewayUdpProxyEnabled: true,
 *     rootCertificateInstallationEnabled: true,
 *     useZtVirtualIp: true,
 * });
 * ```
 *
 * ## Import
 *
 * > This resource does not currently support `pulumi import`.
 */
class ZeroTrustDeviceSettings extends pulumi.CustomResource {
    /**
     * Get an existing ZeroTrustDeviceSettings resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ZeroTrustDeviceSettings(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/zeroTrustDeviceSettings:ZeroTrustDeviceSettings';
    /**
     * Returns true if the given object is an instance of ZeroTrustDeviceSettings.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ZeroTrustDeviceSettings.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["disableForTime"] = state?.disableForTime;
            resourceInputs["externalEmergencySignalEnabled"] = state?.externalEmergencySignalEnabled;
            resourceInputs["externalEmergencySignalFingerprint"] = state?.externalEmergencySignalFingerprint;
            resourceInputs["externalEmergencySignalInterval"] = state?.externalEmergencySignalInterval;
            resourceInputs["externalEmergencySignalUrl"] = state?.externalEmergencySignalUrl;
            resourceInputs["gatewayProxyEnabled"] = state?.gatewayProxyEnabled;
            resourceInputs["gatewayUdpProxyEnabled"] = state?.gatewayUdpProxyEnabled;
            resourceInputs["rootCertificateInstallationEnabled"] = state?.rootCertificateInstallationEnabled;
            resourceInputs["useZtVirtualIp"] = state?.useZtVirtualIp;
        }
        else {
            const args = argsOrState;
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["disableForTime"] = args?.disableForTime;
            resourceInputs["externalEmergencySignalEnabled"] = args?.externalEmergencySignalEnabled;
            resourceInputs["externalEmergencySignalFingerprint"] = args?.externalEmergencySignalFingerprint;
            resourceInputs["externalEmergencySignalInterval"] = args?.externalEmergencySignalInterval;
            resourceInputs["externalEmergencySignalUrl"] = args?.externalEmergencySignalUrl;
            resourceInputs["gatewayProxyEnabled"] = args?.gatewayProxyEnabled;
            resourceInputs["gatewayUdpProxyEnabled"] = args?.gatewayUdpProxyEnabled;
            resourceInputs["rootCertificateInstallationEnabled"] = args?.rootCertificateInstallationEnabled;
            resourceInputs["useZtVirtualIp"] = args?.useZtVirtualIp;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(ZeroTrustDeviceSettings.__pulumiType, name, resourceInputs, opts);
    }
}
exports.ZeroTrustDeviceSettings = ZeroTrustDeviceSettings;
