// @pulumi/cloudflare/zeroTrustNetworkHostnameRoute.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZeroTrustNetworkHostnameRoute = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Cloudflare One Networks Read`
 * - `Cloudflare One Networks Write`
 * - `Cloudflare Tunnel Read`
 * - `Cloudflare Tunnel Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleZeroTrustNetworkHostnameRoute = new cloudflare.ZeroTrustNetworkHostnameRoute("example_zero_trust_network_hostname_route", {
 *     accountId: "699d98642c564d2e855e9661899b7252",
 *     comment: "example comment",
 *     hostname: "office-1.local",
 *     tunnelId: "f70ff985-a4ef-4643-bbbc-4a0ed4fc8415",
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/zeroTrustNetworkHostnameRoute:ZeroTrustNetworkHostnameRoute example '<account_id>/<hostname_route_id>'
 * ```
 */
class ZeroTrustNetworkHostnameRoute extends pulumi.CustomResource {
    /**
     * Get an existing ZeroTrustNetworkHostnameRoute resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ZeroTrustNetworkHostnameRoute(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/zeroTrustNetworkHostnameRoute:ZeroTrustNetworkHostnameRoute';
    /**
     * Returns true if the given object is an instance of ZeroTrustNetworkHostnameRoute.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ZeroTrustNetworkHostnameRoute.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["accountId"] = state?.accountId;
            resourceInputs["comment"] = state?.comment;
            resourceInputs["createdAt"] = state?.createdAt;
            resourceInputs["deletedAt"] = state?.deletedAt;
            resourceInputs["hostname"] = state?.hostname;
            resourceInputs["tunnelId"] = state?.tunnelId;
            resourceInputs["tunnelName"] = state?.tunnelName;
        }
        else {
            const args = argsOrState;
            resourceInputs["accountId"] = args?.accountId;
            resourceInputs["comment"] = args?.comment;
            resourceInputs["hostname"] = args?.hostname;
            resourceInputs["tunnelId"] = args?.tunnelId;
            resourceInputs["createdAt"] = undefined /*out*/;
            resourceInputs["deletedAt"] = undefined /*out*/;
            resourceInputs["tunnelName"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(ZeroTrustNetworkHostnameRoute.__pulumiType, name, resourceInputs, opts);
    }
}
exports.ZeroTrustNetworkHostnameRoute = ZeroTrustNetworkHostnameRoute;
