// @pulumi/cloudflare/zoneDnsSettings.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZoneDnsSettings = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `DNS Read`
 * - `DNS Write`
 * - `Zone DNS Settings Read`
 * - `Zone DNS Settings Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleZoneDnsSettings = new cloudflare.ZoneDnsSettings("example_zone_dns_settings", {
 *     zoneId: "023e105f4ecef8ad9ca31a8372d0c353",
 *     flattenAllCnames: false,
 *     foundationDns: false,
 *     internalDns: {
 *         referenceZoneId: "reference_zone_id",
 *     },
 *     multiProvider: false,
 *     nameservers: {
 *         nsSet: 1,
 *         type: "cloudflare.standard",
 *     },
 *     nsTtl: 86400,
 *     secondaryOverrides: false,
 *     soa: {
 *         expire: 604800,
 *         minTtl: 1800,
 *         mname: "kristina.ns.cloudflare.com",
 *         refresh: 10000,
 *         retry: 2400,
 *         rname: "admin.example.com",
 *         ttl: 3600,
 *     },
 *     zoneMode: "dns_only",
 * });
 * ```
 *
 * ## Import
 *
 * > This resource does not currently support `pulumi import`.
 */
class ZoneDnsSettings extends pulumi.CustomResource {
    /**
     * Get an existing ZoneDnsSettings resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ZoneDnsSettings(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/zoneDnsSettings:ZoneDnsSettings';
    /**
     * Returns true if the given object is an instance of ZoneDnsSettings.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ZoneDnsSettings.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["flattenAllCnames"] = state?.flattenAllCnames;
            resourceInputs["foundationDns"] = state?.foundationDns;
            resourceInputs["internalDns"] = state?.internalDns;
            resourceInputs["multiProvider"] = state?.multiProvider;
            resourceInputs["nameservers"] = state?.nameservers;
            resourceInputs["nsTtl"] = state?.nsTtl;
            resourceInputs["secondaryOverrides"] = state?.secondaryOverrides;
            resourceInputs["soa"] = state?.soa;
            resourceInputs["zoneId"] = state?.zoneId;
            resourceInputs["zoneMode"] = state?.zoneMode;
        }
        else {
            const args = argsOrState;
            resourceInputs["flattenAllCnames"] = args?.flattenAllCnames;
            resourceInputs["foundationDns"] = args?.foundationDns;
            resourceInputs["internalDns"] = args?.internalDns;
            resourceInputs["multiProvider"] = args?.multiProvider;
            resourceInputs["nameservers"] = args?.nameservers;
            resourceInputs["nsTtl"] = args?.nsTtl;
            resourceInputs["secondaryOverrides"] = args?.secondaryOverrides;
            resourceInputs["soa"] = args?.soa;
            resourceInputs["zoneId"] = args?.zoneId;
            resourceInputs["zoneMode"] = args?.zoneMode;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(ZoneDnsSettings.__pulumiType, name, resourceInputs, opts);
    }
}
exports.ZoneDnsSettings = ZoneDnsSettings;
