// @pulumi/cloudflare/zone.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Zone = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `Access: Apps and Policies Read`
 * - `Access: Apps and Policies Revoke`
 * - `Access: Apps and Policies Write`
 * - `Access: Mutual TLS Certificates Write`
 * - `Access: Organizations, Identity Providers, and Groups Write`
 * - `Analytics Read`
 * - `Apps Write`
 * - `Cache Purge`
 * - `DNS Read`
 * - `DNS Write`
 * - `Firewall Services Read`
 * - `Firewall Services Write`
 * - `Load Balancers Read`
 * - `Load Balancers Write`
 * - `Logs Read`
 * - `Logs Write`
 * - `Page Rules Read`
 * - `Page Rules Write`
 * - `SSL and Certificates Read`
 * - `SSL and Certificates Write`
 * - `Stream Read`
 * - `Stream Write`
 * - `Trust and Safety Read`
 * - `Trust and Safety Write`
 * - `Workers Routes Read`
 * - `Workers Routes Write`
 * - `Workers Scripts Read`
 * - `Workers Scripts Write`
 * - `Zaraz Admin`
 * - `Zaraz Edit`
 * - `Zaraz Read`
 * - `Zero Trust: PII Read`
 * - `Zone DNS Edit`
 * - `Zone Read`
 * - `Zone Settings Read`
 * - `Zone Settings Write`
 * - `Zone Write`
 * - `Zone Zone Edit`
 *
 * > If you are attempting to sign up a subdomain of a zone you must first have Subdomain Support entitlement for your account.
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleZone = new cloudflare.Zone("example_zone", {
 *     account: {
 *         id: "023e105f4ecef8ad9ca31a8372d0c353",
 *     },
 *     name: "example.com",
 *     type: "full",
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/zone:Zone example '<zone_id>'
 * ```
 */
class Zone extends pulumi.CustomResource {
    /**
     * Get an existing Zone resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Zone(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/zone:Zone';
    /**
     * Returns true if the given object is an instance of Zone.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Zone.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["account"] = state?.account;
            resourceInputs["activatedOn"] = state?.activatedOn;
            resourceInputs["cnameSuffix"] = state?.cnameSuffix;
            resourceInputs["createdOn"] = state?.createdOn;
            resourceInputs["developmentMode"] = state?.developmentMode;
            resourceInputs["meta"] = state?.meta;
            resourceInputs["modifiedOn"] = state?.modifiedOn;
            resourceInputs["name"] = state?.name;
            resourceInputs["nameServers"] = state?.nameServers;
            resourceInputs["originalDnshost"] = state?.originalDnshost;
            resourceInputs["originalNameServers"] = state?.originalNameServers;
            resourceInputs["originalRegistrar"] = state?.originalRegistrar;
            resourceInputs["owner"] = state?.owner;
            resourceInputs["paused"] = state?.paused;
            resourceInputs["permissions"] = state?.permissions;
            resourceInputs["plan"] = state?.plan;
            resourceInputs["status"] = state?.status;
            resourceInputs["tenant"] = state?.tenant;
            resourceInputs["tenantUnit"] = state?.tenantUnit;
            resourceInputs["type"] = state?.type;
            resourceInputs["vanityNameServers"] = state?.vanityNameServers;
            resourceInputs["verificationKey"] = state?.verificationKey;
        }
        else {
            const args = argsOrState;
            if (args?.account === undefined && !opts.urn) {
                throw new Error("Missing required property 'account'");
            }
            if (args?.name === undefined && !opts.urn) {
                throw new Error("Missing required property 'name'");
            }
            resourceInputs["account"] = args?.account;
            resourceInputs["name"] = args?.name;
            resourceInputs["paused"] = args?.paused;
            resourceInputs["type"] = args?.type;
            resourceInputs["vanityNameServers"] = args?.vanityNameServers;
            resourceInputs["activatedOn"] = undefined /*out*/;
            resourceInputs["cnameSuffix"] = undefined /*out*/;
            resourceInputs["createdOn"] = undefined /*out*/;
            resourceInputs["developmentMode"] = undefined /*out*/;
            resourceInputs["meta"] = undefined /*out*/;
            resourceInputs["modifiedOn"] = undefined /*out*/;
            resourceInputs["nameServers"] = undefined /*out*/;
            resourceInputs["originalDnshost"] = undefined /*out*/;
            resourceInputs["originalNameServers"] = undefined /*out*/;
            resourceInputs["originalRegistrar"] = undefined /*out*/;
            resourceInputs["owner"] = undefined /*out*/;
            resourceInputs["permissions"] = undefined /*out*/;
            resourceInputs["plan"] = undefined /*out*/;
            resourceInputs["status"] = undefined /*out*/;
            resourceInputs["tenant"] = undefined /*out*/;
            resourceInputs["tenantUnit"] = undefined /*out*/;
            resourceInputs["verificationKey"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(Zone.__pulumiType, name, resourceInputs, opts);
    }
}
exports.Zone = Zone;
