// @pulumi/cloudflare/zoneDnssec.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZoneDnssec = void 0;
const pulumi = __importStar(require("@pulumi/pulumi"));
const utilities = __importStar(require("./utilities"));
/**
 * Accepted Permissions
 *
 * - `DNS Read`
 * - `DNS Write`
 *
 * ## Example Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as cloudflare from "@pulumi/cloudflare";
 *
 * const exampleZoneDnssec = new cloudflare.ZoneDnssec("example_zone_dnssec", {
 *     zoneId: "023e105f4ecef8ad9ca31a8372d0c353",
 *     dnssecMultiSigner: false,
 *     dnssecPresigned: true,
 *     dnssecUseNsec3: false,
 *     status: "active",
 * });
 * ```
 *
 * ## Import
 *
 * ```sh
 * $ pulumi import cloudflare:index/zoneDnssec:ZoneDnssec example '<zone_id>'
 * ```
 */
class ZoneDnssec extends pulumi.CustomResource {
    /**
     * Get an existing ZoneDnssec resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new ZoneDnssec(name, state, { ...opts, id: id });
    }
    /** @internal */
    static __pulumiType = 'cloudflare:index/zoneDnssec:ZoneDnssec';
    /**
     * Returns true if the given object is an instance of ZoneDnssec.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ZoneDnssec.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["algorithm"] = state?.algorithm;
            resourceInputs["digest"] = state?.digest;
            resourceInputs["digestAlgorithm"] = state?.digestAlgorithm;
            resourceInputs["digestType"] = state?.digestType;
            resourceInputs["dnssecMultiSigner"] = state?.dnssecMultiSigner;
            resourceInputs["dnssecPresigned"] = state?.dnssecPresigned;
            resourceInputs["dnssecUseNsec3"] = state?.dnssecUseNsec3;
            resourceInputs["ds"] = state?.ds;
            resourceInputs["flags"] = state?.flags;
            resourceInputs["keyTag"] = state?.keyTag;
            resourceInputs["keyType"] = state?.keyType;
            resourceInputs["modifiedOn"] = state?.modifiedOn;
            resourceInputs["publicKey"] = state?.publicKey;
            resourceInputs["status"] = state?.status;
            resourceInputs["zoneId"] = state?.zoneId;
        }
        else {
            const args = argsOrState;
            if (args?.zoneId === undefined && !opts.urn) {
                throw new Error("Missing required property 'zoneId'");
            }
            resourceInputs["dnssecMultiSigner"] = args?.dnssecMultiSigner;
            resourceInputs["dnssecPresigned"] = args?.dnssecPresigned;
            resourceInputs["dnssecUseNsec3"] = args?.dnssecUseNsec3;
            resourceInputs["status"] = args?.status;
            resourceInputs["zoneId"] = args?.zoneId;
            resourceInputs["algorithm"] = undefined /*out*/;
            resourceInputs["digest"] = undefined /*out*/;
            resourceInputs["digestAlgorithm"] = undefined /*out*/;
            resourceInputs["digestType"] = undefined /*out*/;
            resourceInputs["ds"] = undefined /*out*/;
            resourceInputs["flags"] = undefined /*out*/;
            resourceInputs["keyTag"] = undefined /*out*/;
            resourceInputs["keyType"] = undefined /*out*/;
            resourceInputs["modifiedOn"] = undefined /*out*/;
            resourceInputs["publicKey"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(ZoneDnssec.__pulumiType, name, resourceInputs, opts);
    }
}
exports.ZoneDnssec = ZoneDnssec;
