// @tanstack/start-plugin-core/dist/cjs/plugin.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const path = require("node:path");
const nitropack = require("nitropack");
const routerCore = require("@tanstack/router-core");
const startServerCore = require("@tanstack/start-server-core");
const serverFunctionsPlugin = require("@tanstack/server-functions-plugin");
const vite = require("vite");
const vitefu = require("vitefu");
const schema = require("./schema.cjs");
const plugin$5 = require("./nitro-plugin/plugin.cjs");
const plugin$3 = require("./start-manifest-plugin/plugin.cjs");
const startCompilerPlugin = require("./start-compiler-plugin.cjs");
const constants = require("./constants.cjs");
const plugin = require("./start-router-plugin/plugin.cjs");
const plugin$2 = require("./load-env-plugin/plugin.cjs");
const plugin$4 = require("./dev-server-plugin/plugin.cjs");
const plugin$1 = require("./resolve-virtual-entries-plugin/plugin.cjs");
function _interopNamespaceDefault(e) {
  const n = Object.create(null, { [Symbol.toStringTag]: { value: "Module" } });
  if (e) {
    for (const k in e) {
      if (k !== "default") {
        const d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: () => e[k]
        });
      }
    }
  }
  n.default = e;
  return Object.freeze(n);
}
const vite__namespace = /* @__PURE__ */ _interopNamespaceDefault(vite);
schema.createTanStackConfig();
let ssrBundle;
function TanStackStartVitePluginCore(opts, startConfig) {
  return [
    plugin.tanStackStartRouter({
      ...startConfig.tsr,
      target: opts.framework,
      autoCodeSplitting: true
    }),
    plugin$1.resolveVirtualEntriesPlugin(opts, startConfig),
    {
      name: "tanstack-start-core:config-client",
      async config(viteConfig, { command }) {
        var _a;
        const viteAppBase = routerCore.trimPathRight(viteConfig.base || "/");
        globalThis.TSS_APP_BASE = viteAppBase;
        const nitroOutputPublicDir = await (async () => {
          const dummyNitroApp = await nitropack.createNitro({
            preset: startConfig.target,
            compatibilityDate: "2024-12-01"
          });
          const nitroOutputPublicDir2 = dummyNitroApp.options.output.publicDir;
          await dummyNitroApp.close();
          return nitroOutputPublicDir2;
        })();
        const startPackageName = `@tanstack/${opts.framework}-start`;
        const routerPackageName = `@tanstack/${opts.framework}-router`;
        const additionalOptimizeDeps = {
          include: /* @__PURE__ */ new Set(),
          exclude: /* @__PURE__ */ new Set()
        };
        const result = await vitefu.crawlFrameworkPkgs({
          root: process.cwd(),
          isBuild: command === "build",
          isFrameworkPkgByJson(pkgJson) {
            var _a2;
            if ([routerPackageName, startPackageName].includes(pkgJson.name)) {
              return false;
            }
            const peerDependencies = pkgJson["peerDependencies"];
            if (peerDependencies) {
              const internalResult = (_a2 = opts.crawlPackages) == null ? void 0 : _a2.call(opts, {
                name: pkgJson.name,
                peerDependencies,
                exports: pkgJson.exports
              });
              if (internalResult) {
                if (internalResult === "exclude") {
                  additionalOptimizeDeps.exclude.add(pkgJson.name);
                } else {
                  additionalOptimizeDeps.include.add(pkgJson.name);
                }
              }
              return startPackageName in peerDependencies || routerPackageName in peerDependencies;
            }
            return false;
          }
        });
        return {
          base: viteAppBase,
          environments: {
            [constants.VITE_ENVIRONMENT_NAMES.client]: {
              consumer: "client",
              build: {
                manifest: true,
                rollupOptions: {
                  input: {
                    main: getClientEntryPath(startConfig)
                  },
                  output: {
                    dir: path.resolve(startConfig.root, constants.CLIENT_DIST_DIR)
                  },
                  // TODO: this should be removed
                  external: ["node:fs", "node:path", "node:os", "node:crypto"]
                }
              }
            },
            [constants.VITE_ENVIRONMENT_NAMES.server]: {
              consumer: "server",
              build: {
                ssr: true,
                // we don't write to the file system as the below 'capture-output' plugin will
                // capture the output and write it to the virtual file system
                write: false,
                copyPublicDir: false,
                rollupOptions: {
                  output: {
                    entryFileNames: constants.SSR_ENTRY_FILE
                  },
                  plugins: [
                    {
                      name: "capture-output",
                      generateBundle(_options, bundle) {
                        ssrBundle = bundle;
                      }
                    }
                  ]
                },
                commonjsOptions: {
                  include: [/node_modules/]
                }
              }
            }
          },
          resolve: {
            noExternal: [
              "@tanstack/start**",
              `@tanstack/${opts.framework}-start**`,
              ...Object.values(startServerCore.VIRTUAL_MODULES),
              startPackageName,
              ...result.ssr.noExternal.sort()
            ],
            external: [...result.ssr.external.sort()],
            dedupe: [startPackageName]
          },
          optimizeDeps: {
            exclude: [
              ...Object.values(startServerCore.VIRTUAL_MODULES),
              startPackageName,
              ...result.optimizeDeps.exclude.sort(),
              ...additionalOptimizeDeps.exclude
            ],
            include: [...additionalOptimizeDeps.include]
          },
          /* prettier-ignore */
          define: {
            // define is an esbuild function that replaces the any instances of given keys with the given values
            // i.e: __FRAMEWORK_NAME__ can be replaced with JSON.stringify("TanStack Start")
            // This is not the same as injecting environment variables.
            ...defineReplaceEnv("TSS_SERVER_FN_BASE", startConfig.serverFns.base),
            ...defineReplaceEnv("TSS_OUTPUT_PUBLIC_DIR", nitroOutputPublicDir),
            ...defineReplaceEnv("TSS_APP_BASE", viteAppBase),
            ...command === "serve" ? defineReplaceEnv("TSS_SHELL", ((_a = startConfig.spa) == null ? void 0 : _a.enabled) ? "true" : "false") : {}
          }
        };
      }
    },
    // N.B. TanStackStartCompilerPlugin must be before the TanStackServerFnPluginEnv
    startCompilerPlugin.startCompilerPlugin(opts.framework, {
      client: { envName: constants.VITE_ENVIRONMENT_NAMES.client },
      server: { envName: constants.VITE_ENVIRONMENT_NAMES.server }
    }),
    serverFunctionsPlugin.TanStackServerFnPluginEnv({
      // This is the ID that will be available to look up and import
      // our server function manifest and resolve its module
      manifestVirtualImportId: startServerCore.VIRTUAL_MODULES.serverFnManifest,
      client: {
        getRuntimeCode: () => `import { createClientRpc } from '@tanstack/${opts.framework}-start/server-functions-client'`,
        replacer: (d) => `createClientRpc('${d.functionId}', '${startConfig.serverFns.base}')`,
        envName: constants.VITE_ENVIRONMENT_NAMES.client
      },
      server: {
        getRuntimeCode: () => `import { createServerRpc } from '@tanstack/${opts.framework}-start/server-functions-server'`,
        replacer: (d) => `createServerRpc('${d.functionId}', '${startConfig.serverFns.base}', ${d.fn})`,
        envName: constants.VITE_ENVIRONMENT_NAMES.server
      }
    }),
    plugin$2.loadEnvPlugin(startConfig),
    plugin$3.startManifestPlugin({ clientEntry: getClientEntryPath(startConfig) }),
    plugin$4.devServerPlugin(),
    plugin$5.nitroPlugin(startConfig, () => ssrBundle),
    {
      name: "tanstack-start:core:capture-client-bundle",
      applyToEnvironment(e) {
        return e.name === constants.VITE_ENVIRONMENT_NAMES.client;
      },
      enforce: "post",
      generateBundle(_options, bundle) {
        globalThis.TSS_CLIENT_BUNDLE = bundle;
      }
    }
  ];
}
function defineReplaceEnv(key, value) {
  return {
    [`process.env.${key}`]: JSON.stringify(value),
    [`import.meta.env.${key}`]: JSON.stringify(value)
  };
}
const getClientEntryPath = (startConfig) => {
  const entry = startConfig.clientEntryPath.startsWith(
    "/~start/default-client-entry"
  ) ? startConfig.clientEntryPath : vite__namespace.normalizePath(
    path.join(
      "/@fs",
      path.resolve(startConfig.root, startConfig.clientEntryPath)
    )
  );
  return entry;
};
exports.TanStackStartVitePluginCore = TanStackStartVitePluginCore;
