// @trigger.dev/core/dist/commonjs/v3/jwt.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JWT_AUDIENCE = exports.JWT_ISSUER = exports.JWT_ALGORITHM = void 0;
exports.generateJWT = generateJWT;
exports.validateJWT = validateJWT;
exports.JWT_ALGORITHM = "HS256";
exports.JWT_ISSUER = "https://id.trigger.dev";
exports.JWT_AUDIENCE = "https://api.trigger.dev";
async function generateJWT(options) {
    const { SignJWT } = await import("jose");
    const secret = new TextEncoder().encode(options.secretKey);
    return new SignJWT(options.payload)
        .setIssuer(exports.JWT_ISSUER)
        .setAudience(exports.JWT_AUDIENCE)
        .setProtectedHeader({ alg: exports.JWT_ALGORITHM })
        .setIssuedAt()
        .setExpirationTime(options.expirationTime ?? "15m")
        .sign(secret);
}
async function validateJWT(token, apiKey) {
    const { jwtVerify, errors } = await import("jose");
    const secret = new TextEncoder().encode(apiKey);
    try {
        const { payload } = await jwtVerify(token, secret, {
            issuer: exports.JWT_ISSUER,
            audience: exports.JWT_AUDIENCE,
        });
        return { ok: true, payload };
    }
    catch (error) {
        if (error instanceof errors.JOSEError) {
            return {
                ok: false,
                error: error.message,
                code: error.code,
            };
        }
        else {
            return {
                ok: false,
                error: error instanceof Error ? error.message : "Unknown error",
                code: "ERR_UNKNOWN",
            };
        }
    }
}
