// @tanstack/start-plugin-core/dist/cjs/nitro-plugin/plugin.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const path = require("node:path");
const node_fs = require("node:fs");
const nitropack = require("nitropack");
const pathe = require("pathe");
const startServerCore = require("@tanstack/start-server-core");
const constants = require("../constants.cjs");
const buildSitemap = require("./build-sitemap.cjs");
const prerender = require("./prerender.cjs");
function nitroPlugin(options, getSsrBundle) {
  const buildPreset = process.env["START_TARGET"] ?? options.target;
  return [
    {
      name: "tanstack-vite-plugin-nitro",
      configEnvironment(name) {
        if (name === constants.VITE_ENVIRONMENT_NAMES.server) {
          return {
            build: {
              commonjsOptions: {
                include: []
              },
              ssr: true,
              sourcemap: true,
              rollupOptions: {
                input: "/~start/server-entry"
              }
            }
          };
        }
        return null;
      },
      config() {
        return {
          builder: {
            sharedPlugins: true,
            async buildApp(builder) {
              const client = builder.environments[constants.VITE_ENVIRONMENT_NAMES.client];
              const server = builder.environments[constants.VITE_ENVIRONMENT_NAMES.server];
              if (!client) {
                throw new Error("Client environment not found");
              }
              if (!server) {
                throw new Error("SSR environment not found");
              }
              const clientOutputDir = pathe.resolve(options.root, constants.CLIENT_DIST_DIR);
              node_fs.rmSync(clientOutputDir, { recursive: true, force: true });
              await builder.build(client);
              await builder.build(server);
              const nitroConfig = {
                dev: false,
                // TODO: do we need this? should this be made configurable?
                compatibilityDate: "2024-11-19",
                logLevel: 3,
                preset: buildPreset,
                baseURL: globalThis.TSS_APP_BASE,
                publicAssets: [
                  {
                    dir: path.resolve(options.root, constants.CLIENT_DIST_DIR),
                    baseURL: "/",
                    maxAge: 31536e3
                    // 1 year
                  }
                ],
                typescript: {
                  generateTsConfig: false
                },
                prerender: void 0,
                renderer: constants.SSR_ENTRY_FILE,
                plugins: [],
                // Nitro's plugins
                appConfigFiles: [],
                scanDirs: [],
                imports: false,
                // unjs/unimport for global/magic imports
                rollupConfig: {
                  plugins: [virtualBundlePlugin(getSsrBundle())]
                },
                virtual: {
                  // This is Nitro's way of defining virtual modules
                  // Should we define the ones for TanStack Start's here as well?
                }
              };
              const nitro = await nitropack.createNitro(nitroConfig);
              await buildNitroApp(builder, nitro, options);
            }
          }
        };
      }
    }
  ];
}
async function buildNitroApp(builder, nitro, options) {
  var _a, _b, _c;
  await nitropack.prepare(nitro);
  await nitropack.copyPublicAssets(nitro);
  if (((_a = options.prerender) == null ? void 0 : _a.enabled) !== false) {
    options.prerender = {
      ...options.prerender,
      enabled: ((_b = options.prerender) == null ? void 0 : _b.enabled) ?? options.pages.some(
        (d) => {
          var _a2;
          return typeof d === "string" ? false : !!((_a2 = d.prerender) == null ? void 0 : _a2.enabled);
        }
      )
    };
  }
  if ((_c = options.spa) == null ? void 0 : _c.enabled) {
    options.prerender = {
      ...options.prerender,
      enabled: true
    };
    const maskUrl = new URL(options.spa.maskPath, "http://localhost");
    options.pages.push({
      path: maskUrl.toString().replace("http://localhost", ""),
      prerender: {
        ...options.spa.prerender,
        headers: {
          ...options.spa.prerender.headers,
          [startServerCore.HEADERS.TSS_SHELL]: "true"
        }
      },
      sitemap: {
        exclude: true
      }
    });
  }
  if (options.prerender.enabled) {
    await prerender.prerender({
      options,
      nitro,
      builder
    });
  }
  if (options.pages.length) {
    buildSitemap.buildSitemap({
      options,
      publicDir: nitro.options.output.publicDir
    });
  }
  await nitropack.build(nitro);
  await nitro.close();
  nitro.logger.success(
    "Client and Server bundles for TanStack Start have been successfully built."
  );
}
function virtualBundlePlugin(ssrBundle) {
  const _modules = /* @__PURE__ */ new Map();
  for (const [fileName, content] of Object.entries(ssrBundle)) {
    if (content.type === "chunk") {
      const virtualModule = {
        code: content.code,
        map: null
      };
      const maybeMap = ssrBundle[`${fileName}.map`];
      if (maybeMap && maybeMap.type === "asset") {
        virtualModule.map = maybeMap.source;
      }
      _modules.set(fileName, virtualModule);
      _modules.set(pathe.resolve(fileName), virtualModule);
    }
  }
  return {
    name: "virtual-bundle",
    resolveId(id, importer) {
      if (_modules.has(id)) {
        return pathe.resolve(id);
      }
      if (importer) {
        const resolved = pathe.resolve(pathe.dirname(importer), id);
        if (_modules.has(resolved)) {
          return resolved;
        }
      }
      return null;
    },
    load(id) {
      const m = _modules.get(id);
      if (!m) {
        return null;
      }
      return m;
    }
  };
}
exports.nitroPlugin = nitroPlugin;
