// @trigger.dev/core/dist/esm/v3/runTimelineMetrics/runTimelineMetricsManager.js
import { getEnvVar } from "../utils/getEnv.js";
export class StandardRunTimelineMetricsManager {
    _metrics = [];
    registerMetric(metric) {
        this._metrics.push(metric);
    }
    getMetrics() {
        return this._metrics;
    }
    registerMetricsFromExecution(metrics) {
        if (metrics) {
            metrics.forEach((metric) => {
                this.registerMetric({
                    name: `trigger.dev/${metric.name}`,
                    event: metric.event,
                    timestamp: metric.timestamp,
                    attributes: {
                        duration: metric.duration,
                    },
                });
            });
        }
    }
    seedMetricsFromEnvironment() {
        const forkStartTime = getEnvVar("TRIGGER_PROCESS_FORK_START_TIME");
        if (typeof forkStartTime === "string") {
            const forkStartTimeMs = parseInt(forkStartTime, 10);
            this.registerMetric({
                name: "trigger.dev/start",
                event: "fork",
                attributes: {
                    duration: Date.now() - forkStartTimeMs,
                },
                timestamp: forkStartTimeMs,
            });
        }
    }
}
export class NoopRunTimelineMetricsManager {
    registerMetric(metric) {
        // Do nothing
    }
    getMetrics() {
        return [];
    }
}
